// @xviz/builder/dist/esm/builders/declarative-ui/xviz-ui-builder.js
var _UI_BUILDER_MAP;

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/* global console */

/* eslint-disable no-console */
import XVIZPanelBuilder from './xviz-panel-builder';
import XVIZContainerBuilder from './xviz-container-builder';
import XVIZMetricBuilder from './xviz-metric-builder';
import XVIZPlotBuilder from './xviz-plot-builder';
import XVIZSelectBuilder from './xviz-select-builder';
import XVIZTableBuilder from './xviz-table-builder';
import XVIZTreeTableBuilder from './xviz-tree-table-builder';
import XVIZVideoBuilder from './xviz-video-builder';
import { snakeToCamel } from './utils';
import { UI_TYPES } from './constants';
var defaultValidateWarn = console.warn;
var defaultValidateError = console.error;
var UI_BUILDER_MAP = (_UI_BUILDER_MAP = {}, _defineProperty(_UI_BUILDER_MAP, UI_TYPES.PANEL, XVIZPanelBuilder), _defineProperty(_UI_BUILDER_MAP, UI_TYPES.CONTAINER, XVIZContainerBuilder), _defineProperty(_UI_BUILDER_MAP, UI_TYPES.METRIC, XVIZMetricBuilder), _defineProperty(_UI_BUILDER_MAP, UI_TYPES.PLOT, XVIZPlotBuilder), _defineProperty(_UI_BUILDER_MAP, UI_TYPES.SELECT, XVIZSelectBuilder), _defineProperty(_UI_BUILDER_MAP, UI_TYPES.TABLE, XVIZTableBuilder), _defineProperty(_UI_BUILDER_MAP, UI_TYPES.TREETABLE, XVIZTreeTableBuilder), _defineProperty(_UI_BUILDER_MAP, UI_TYPES.VIDEO, XVIZVideoBuilder), _UI_BUILDER_MAP);

var XVIZUIBuilder =
/*#__PURE__*/
function () {
  function XVIZUIBuilder() {
    var _this = this;

    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    _classCallCheck(this, XVIZUIBuilder);

    this._validateWarn = options.validateWarn || defaultValidateWarn;
    this._validateError = options.validateError || defaultValidateError;
    this._children = [];
    Object.values(UI_TYPES).map(function (type) {
      // add UI builders, e.g.
      // type `panel`
      //  - this.panel = (props) => this._setChild('panel', props);
      var camelType = snakeToCamel(type);

      _this[camelType] = function (props) {
        return _this._createUIBuilder(type, props);
      };
    });
  }

  _createClass(XVIZUIBuilder, [{
    key: "getUI",
    value: function getUI() {
      return this._children.reduce(function (ui, child) {
        var childConfig = child.getUI();
        ui[childConfig.name] = childConfig;
        return ui;
      }, {});
    }
  }, {
    key: "child",
    value: function child(_child) {
      if (!(_child instanceof XVIZPanelBuilder)) {
        this._validateError('Top level UI element should be `Panel`');
      }

      this._children.push(_child);

      return _child;
    }
  }, {
    key: "_createUIBuilder",
    value: function _createUIBuilder(type, props) {
      return new UI_BUILDER_MAP[type](_objectSpread({}, props, {
        validateWarn: this._validateWarn,
        validateError: this._validateError,
        root: this
      }));
    }
  }]);

  return XVIZUIBuilder;
}();

export { XVIZUIBuilder as default };
