// @xviz/builder/dist/es5/builders/xviz-builder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _xvizPoseBuilder = _interopRequireDefault(require("./xviz-pose-builder"));

var _xvizPrimitiveBuilder = _interopRequireDefault(require("./xviz-primitive-builder"));

var _xvizFutureInstanceBuilder = _interopRequireDefault(require("./xviz-future-instance-builder"));

var _xvizUiPrimitiveBuilder = _interopRequireDefault(require("./xviz-ui-primitive-builder"));

var _xvizTimeSeriesBuilder = _interopRequireDefault(require("./xviz-time-series-builder"));

var _xvizValidator = _interopRequireDefault(require("./xviz-validator"));

var _xvizVariableBuilder = _interopRequireDefault(require("./xviz-variable-builder"));

var _constant = require("./constant");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

/* global console */

/* eslint-disable no-console */
var defaultValidateWarn = console.warn;
var defaultValidateError = console.error;
/* eslint-enable no-console */
// TODO: Builder could validate against stream metadata!

var XVIZBuilder =
/*#__PURE__*/
function () {
  function XVIZBuilder() {
    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
        _ref$metadata = _ref.metadata,
        metadata = _ref$metadata === void 0 ? {} : _ref$metadata,
        _ref$disableStreams = _ref.disableStreams,
        disableStreams = _ref$disableStreams === void 0 ? [] : _ref$disableStreams,
        _ref$validateWarn = _ref.validateWarn,
        validateWarn = _ref$validateWarn === void 0 ? defaultValidateWarn : _ref$validateWarn,
        _ref$validateError = _ref.validateError,
        validateError = _ref$validateError === void 0 ? defaultValidateError : _ref$validateError;

    _classCallCheck(this, XVIZBuilder);

    this._validator = new _xvizValidator.default({
      validateWarn: validateWarn,
      validateError: validateError
    });
    this.metadata = metadata;
    this.disableStreams = disableStreams; // Current streamBuilder

    this._streamBuilder = null; // Construct different builders

    this._poseBuilder = new _xvizPoseBuilder.default({
      metadata: this.metadata,
      validator: this._validator
    });
    this._variablesBuilder = new _xvizVariableBuilder.default({
      metadata: this.metadata,
      validator: this._validator
    });
    this._primitivesBuilder = new _xvizPrimitiveBuilder.default({
      metadata: this.metadata,
      validator: this._validator
    });
    this._futureInstanceBuilder = new _xvizFutureInstanceBuilder.default({
      metadata: this.metadata,
      validator: this._validator
    });
    this._uiPrimitivesBuilder = new _xvizUiPrimitiveBuilder.default({
      metadata: this.metadata,
      validator: this._validator
    });
    this._timeSeriesBuilder = new _xvizTimeSeriesBuilder.default({
      metadata: this.metadata,
      validator: this._validator
    });
  }

  _createClass(XVIZBuilder, [{
    key: "pose",
    value: function pose() {
      var streamId = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _constant.PRIMARY_POSE_STREAM;
      this._streamBuilder = this._poseBuilder.stream(streamId);
      return this._streamBuilder;
    }
  }, {
    key: "variable",
    value: function variable(streamId) {
      this._streamBuilder = this._variablesBuilder.stream(streamId);
      return this._streamBuilder;
    }
  }, {
    key: "primitive",
    value: function primitive(streamId) {
      this._streamBuilder = this._primitivesBuilder.stream(streamId);
      return this._streamBuilder;
    }
  }, {
    key: "futureInstance",
    value: function futureInstance(streamId, timestamp) {
      this._streamBuilder = this._futureInstanceBuilder.stream(streamId);

      this._streamBuilder._timestamp(timestamp);

      return this._streamBuilder;
    }
  }, {
    key: "uiPrimitive",
    value: function uiPrimitive(streamId) {
      this._streamBuilder = this._uiPrimitivesBuilder.stream(streamId);
      return this._streamBuilder;
    }
  }, {
    key: "timeSeries",
    value: function timeSeries(streamId) {
      this._streamBuilder = this._timeSeriesBuilder.stream(streamId);
      return this._streamBuilder;
    }
    /*
    frame data:
    {
      update_type: 'snapshot',
      updates: [{
        timestamp,
        poses: {'/vehicle-pose': {}, ...},
        primitives: {},
        variables: {},
        future_instances: {}
      }]
    }
     */

  }, {
    key: "getFrame",
    value: function getFrame() {
      var _this$_poseBuilder$ge = this._poseBuilder.getData(),
          poses = _this$_poseBuilder$ge.poses;

      if (!poses || !poses[_constant.PRIMARY_POSE_STREAM]) {
        this._validator.error("Every frame requires a ".concat(_constant.PRIMARY_POSE_STREAM, " stream"));
      }

      var primitives = this._primitivesBuilder.getData();

      var futures = this._futureInstanceBuilder.getData();

      var variables = this._variablesBuilder.getData();

      var timeSeries = this._timeSeriesBuilder.getData();

      var uiPrimitives = this._uiPrimitivesBuilder.getData();

      var data = {
        timestamp: poses[_constant.PRIMARY_POSE_STREAM].timestamp,
        poses: poses
      };

      if (primitives) {
        data.primitives = primitives;
      }

      if (futures) {
        data.future_instances = futures;
      }

      if (variables) {
        data.variables = variables;
      }

      if (timeSeries) {
        data.time_series = timeSeries;
      }

      if (uiPrimitives) {
        data.ui_primitives = uiPrimitives;
      }

      var frame = {
        update_type: 'snapshot',
        updates: [data]
      };
      return frame;
    }
  }, {
    key: "_reset",
    value: function _reset() {
      this._streamBuilder = null;
    }
  }]);

  return XVIZBuilder;
}();

exports.default = XVIZBuilder;
