// @streetscape.gl/monochrome/dist/esm/shared/slider/index.js
function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { withTheme, evaluateStyle } from '../theme';
import Draggable from '../draggable';
import { clamp } from '../../utils/math';

function snap(x, min, max, step) {
  if (step > 0) {
    x = Math.round((x - min) / step) * step + min;
  }

  return clamp(x, min, max);
}

var SliderWrapper = styled.div(function (props) {
  return _objectSpread({}, props.theme.__reset__, {
    outline: 'none',
    color: props.isEnabled ? props.theme.textColorPrimary : props.theme.textColorDisabled,
    cursor: 'pointer',
    pointerEvents: props.isEnabled ? 'all' : 'none',
    paddingTop: props.knobSize / 2,
    paddingBottom: props.knobSize / 2
  }, evaluateStyle(props.userStyle, props));
});
var SliderTrack = styled.div(function (props) {
  return _objectSpread({
    position: 'relative',
    width: '100%',
    background: props.isEnabled ? props.theme.controlColorPrimary : props.theme.controlColorDisabled,
    height: 2
  }, evaluateStyle(props.userStyle, props));
});
var SliderTrackFill = styled.div(function (props) {
  return _objectSpread({
    position: 'absolute',
    transitionProperty: 'width',
    transitionDuration: props.isDragging ? '0s' : props.theme.transitionDuration,
    transitionTimingFunction: props.theme.transitionTimingFunction,
    height: '100%',
    background: props.isEnabled ? props.theme.controlColorActive : props.theme.controlColorDisabled
  }, evaluateStyle(props.userStyle, props));
});
var SliderKnob = styled.div(function (props) {
  return _objectSpread({
    position: 'absolute',
    borderStyle: 'solid',
    borderWidth: 2,
    borderColor: props.isEnabled ? props.isHovered ? props.theme.controlColorHovered : props.hasFocus ? props.theme.controlColorActive : props.theme.controlColorPrimary : props.theme.controlColorDisabled,
    background: props.theme.background,
    boxSizing: 'border-box',
    width: props.knobSize,
    height: props.knobSize,
    borderRadius: '50%',
    margin: -props.knobSize / 2,
    top: '50%',
    transitionProperty: 'left',
    transitionDuration: props.isDragging ? '0s' : props.theme.transitionDuration
  }, evaluateStyle(props.userStyle, props));
});
/*
 * @class
 */

var Slider =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(Slider, _PureComponent);

  function Slider() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, Slider);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(Slider)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "state", {
      width: 1,
      isHovered: false,
      hasFocus: false,
      isDragging: false,
      hasDragged: false
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_updateValue", function (offsetX, width) {
      var _this$props = _this.props,
          min = _this$props.min,
          max = _this$props.max,
          step = _this$props.step;
      var pos = clamp(offsetX / width, 0, 1);
      var value = snap(min + (max - min) * pos, min, max, step);

      _this.props.onChange(value);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onMouseEnter", function () {
      return _this.setState({
        isHovered: true
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onMouseLeave", function () {
      return _this.setState({
        isHovered: false
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onFocus", function () {
      return _this.setState({
        hasFocus: true
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onBlur", function () {
      return _this.setState({
        hasFocus: false
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onKeyDown", function (evt) {
      var delta;

      switch (evt.keyCode) {
        case 37:
          // left
          delta = -1;
          break;

        case 39:
          // right
          delta = 1;
          break;

        default:
          return;
      }

      var _this$props2 = _this.props,
          value = _this$props2.value,
          min = _this$props2.min,
          max = _this$props2.max;
      var step = _this.props.step || (max - min) / 20;
      var newValue = clamp(value + step * delta, min, max);

      if (newValue !== value) {
        _this.props.onChange(newValue);
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onDragStart", function (evt) {
      var width = _this._track.clientWidth;

      _this.setState({
        width: width
      });

      _this._updateValue(evt.offsetX, width);

      _this.setState({
        isDragging: true
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onDrag", function (evt) {
      _this._updateValue(evt.offsetX, _this.state.width);

      _this.setState({
        hasDragged: evt.hasDragged
      });
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onDragEnd", function (evt) {
      _this.setState({
        isDragging: false,
        hasDragged: false
      });
    });

    return _this;
  }

  _createClass(Slider, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props3 = this.props,
          value = _this$props3.value,
          min = _this$props3.min,
          max = _this$props3.max,
          step = _this$props3.step,
          isEnabled = _this$props3.isEnabled,
          children = _this$props3.children,
          className = _this$props3.className,
          style = _this$props3.style,
          theme = _this$props3.theme;
      var _this$state = this.state,
          isHovered = _this$state.isHovered,
          isDragging = _this$state.isDragging,
          hasFocus = _this$state.hasFocus,
          hasDragged = _this$state.hasDragged;
      var _style$tolerance = style.tolerance,
          tolerance = _style$tolerance === void 0 ? 0 : _style$tolerance,
          _style$knobSize = style.knobSize,
          knobSize = _style$knobSize === void 0 ? theme.controlSize : _style$knobSize;
      var ratio = (snap(value, min, max, step) - min) / (max - min);
      var styleProps = {
        theme: theme,
        knobSize: knobSize,
        isEnabled: isEnabled,
        isHovered: isHovered,
        hasFocus: hasFocus,
        isActive: isDragging,
        isDragging: hasDragged
      };
      return React.createElement(SliderWrapper, _extends({}, styleProps, {
        userStyle: style.wrapper,
        className: className,
        tabIndex: isEnabled ? 0 : -1,
        onMouseEnter: this._onMouseEnter,
        onMouseLeave: this._onMouseLeave,
        onFocus: this._onFocus,
        onBlur: this._onBlur,
        onKeyDown: this._onKeyDown
      }), React.createElement(Draggable, {
        tolerance: knobSize / 2 + tolerance,
        onDragStart: this._onDragStart,
        onDrag: this._onDrag,
        onDragEnd: this._onDragEnd
      }, React.createElement(SliderTrack, _extends({
        userStyle: style.track
      }, styleProps, {
        ref: function ref(_ref) {
          _this2._track = _ref;
        }
      }), children, React.createElement(SliderTrackFill, _extends({}, styleProps, {
        userStyle: style.trackFill,
        style: {
          width: "".concat(ratio * 100, "%")
        }
      })), React.createElement(SliderKnob, _extends({}, styleProps, {
        userStyle: style.knob,
        style: {
          left: "".concat(ratio * 100, "%")
        }
      })))));
    }
  }]);

  return Slider;
}(PureComponent);

_defineProperty(Slider, "propTypes", {
  value: PropTypes.number.isRequired,
  min: PropTypes.number.isRequired,
  max: PropTypes.number.isRequired,
  onChange: PropTypes.func,
  className: PropTypes.string,
  style: PropTypes.object,
  step: PropTypes.number,
  label: PropTypes.string,
  tooltip: PropTypes.string,
  badge: PropTypes.element,
  isEnabled: PropTypes.bool
});

_defineProperty(Slider, "defaultProps", {
  className: '',
  style: {},
  step: 0,
  isEnabled: true,
  onChange: function onChange() {}
});

export default withTheme(Slider);
