// @loaders.gl/gltf/dist/es5/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "GLTFLoader", {
  enumerable: true,
  get: function get() {
    return _gltfLoader.default;
  }
});
Object.defineProperty(exports, "GLTFWriter", {
  enumerable: true,
  get: function get() {
    return _gltfWriter.default;
  }
});
Object.defineProperty(exports, "GLTFParser", {
  enumerable: true,
  get: function get() {
    return _gltfParser.default;
  }
});
Object.defineProperty(exports, "GLTFBuilder", {
  enumerable: true,
  get: function get() {
    return _gltfBuilder.default;
  }
});
Object.defineProperty(exports, "GLBParser", {
  enumerable: true,
  get: function get() {
    return _glbParser.default;
  }
});
Object.defineProperty(exports, "GLBBuilder", {
  enumerable: true,
  get: function get() {
    return _glbBuilder.default;
  }
});
Object.defineProperty(exports, "ATTRIBUTE_TYPE_TO_COMPONENTS", {
  enumerable: true,
  get: function get() {
    return _gltfTypeUtils.ATTRIBUTE_TYPE_TO_COMPONENTS;
  }
});
Object.defineProperty(exports, "ATTRIBUTE_COMPONENT_TYPE_TO_BYTE_SIZE", {
  enumerable: true,
  get: function get() {
    return _gltfTypeUtils.ATTRIBUTE_COMPONENT_TYPE_TO_BYTE_SIZE;
  }
});
Object.defineProperty(exports, "ATTRIBUTE_COMPONENT_TYPE_TO_ARRAY", {
  enumerable: true,
  get: function get() {
    return _gltfTypeUtils.ATTRIBUTE_COMPONENT_TYPE_TO_ARRAY;
  }
});

var _gltfLoader = _interopRequireDefault(require("./gltf/gltf-loader"));

var _gltfWriter = _interopRequireDefault(require("./gltf/gltf-writer"));

var _gltfParser = _interopRequireDefault(require("./gltf/gltf-parser"));

var _gltfBuilder = _interopRequireDefault(require("./gltf/gltf-builder"));

var _glbParser = _interopRequireDefault(require("./glb/glb-parser"));

var _glbBuilder = _interopRequireDefault(require("./glb/glb-builder"));

var _gltfTypeUtils = require("./utils/gltf-type-utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
