// @xviz/builder/dist/es6/writers/xviz-writer/xviz-writer.js
// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { writeBinaryXVIZtoFile } from './xviz-binary-writer';
import { xvizConvertJson } from './xviz-json-encoder.js';
import { DracoEncoder, DracoDecoder } from '@loaders.gl/draco'; // 0-frame is an index file for timestamp metadata
// 1-frame is the metadata file for the log
// 2-frame is where the actual XVIZ updates begin

const frameName = index => `${index + 2}-frame`;
/**
 * Class to abstract away file IO
 */


class FileSink {
  constructor() {
    this.fs = module.require('fs');
    this.path = module.require('path');
  }

  writeSync(scope, name, data) {
    const xvizMetadataFilename = this.path.join(scope, name);
    const options = {
      flag: 'w'
    };
    this.fs.writeFileSync(xvizMetadataFilename, data, options);
  }

}

export default class XVIZWriter {
  constructor() {
    let options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    const _options$dataSink = options.dataSink,
          dataSink = _options$dataSink === void 0 ? new FileSink() : _options$dataSink,
          _options$envelope = options.envelope,
          envelope = _options$envelope === void 0 ? true : _options$envelope,
          _options$binary = options.binary,
          binary = _options$binary === void 0 ? true : _options$binary,
          _options$json = options.json,
          json = _options$json === void 0 ? false : _options$json,
          _options$draco = options.draco,
          draco = _options$draco === void 0 ? false : _options$draco;
    this.sink = dataSink;
    this.frameTimings = {
      frames: new Map()
    };
    this.wroteFrameIndex = null;
    this.options = {
      envelope,
      binary,
      json,
      draco
    };
  } // xvizMetadata is the object returned
  // from a Builder.


  writeMetadata(xvizDirectory, xvizMetadata) {
    this._saveTimestamp(xvizMetadata);

    if (this.options.envelope) {
      xvizMetadata = {
        type: 'xviz/metadata',
        data: xvizMetadata
      };
    }

    if (this.options.binary) {
      const options = {
        flattenArrays: false
      };
      writeBinaryXVIZtoFile(this.sink, xvizDirectory, '1-frame', xvizMetadata, options);
    }

    if (this.options.json) {
      this.sink.writeSync(xvizDirectory, '1-frame.json', JSON.stringify(xvizMetadata));
    }
  }

  writeFrame(xvizDirectory, frameIndex, xvizFrame) {
    if (this.wroteFrameIndex !== null) {
      throw new Error(`writeFrame() was called after writeFrameIndex().  The index was written with last frame of ${frameName(this.wroteFrameIndex - 1)}`);
    }

    this._saveTimestamp(xvizFrame, frameIndex);

    if (this.options.envelope) {
      xvizFrame = {
        type: 'xviz/state_update',
        data: xvizFrame
      };
    }

    if (this.options.binary) {
      const options = {
        flattenArrays: false
      };

      if (this.options.draco) {
        options.DracoEncoder = DracoEncoder;
        options.DracoDecoder = DracoDecoder;
      }

      writeBinaryXVIZtoFile(this.sink, xvizDirectory, frameName(frameIndex), xvizFrame, options);
    }

    if (this.options.json) {
      // Limit precision to save space
      const numberRounder = (k, value) => {
        if (typeof value === 'number') {
          return Number(value.toFixed(10));
        }

        return value;
      };

      const jsonXVIZFrame = xvizConvertJson(xvizFrame);
      this.sink.writeSync(xvizDirectory, `${frameName(frameIndex)}.json`, JSON.stringify(jsonXVIZFrame, numberRounder));
    }
  }

  writeFrameIndex(xvizDirectory) {
    const _this$frameTimings = this.frameTimings,
          startTime = _this$frameTimings.startTime,
          endTime = _this$frameTimings.endTime,
          frames = _this$frameTimings.frames;
    const frameTimings = {};

    if (startTime) {
      frameTimings.startTime = startTime;
    }

    if (endTime) {
      frameTimings.endTime = endTime;
    } // Sort frames by index before writing out as an array


    const frameTimes = Array.from(frames.keys()).sort((a, b) => a - b);
    const timing = [];
    frameTimes.forEach((value, index) => {
      // Value is two greater than frame index
      const limit = timing.length;

      if (value > limit) {
        // Adding 2 because 1-frame is metadata file, so frame data starts at 2
        throw new Error(`Error writing time index file. Frames are missing between ${limit + 2} and ${value + 2}`);
      }

      timing.push(frames.get(value));
    });
    frameTimings.timing = timing;
    this.sink.writeSync(xvizDirectory, '0-frame.json', JSON.stringify(frameTimings));
    this.wroteFrameIndex = timing.length;
  }
  /* eslint-disable camelcase */


  _saveTimestamp(xviz_data, index) {
    const log_info = xviz_data.log_info,
          updates = xviz_data.updates;

    if (index === undefined) {
      // Metadata case
      if (log_info) {
        const _ref = log_info || {},
              start_time = _ref.start_time,
              end_time = _ref.end_time;

        if (start_time) {
          this.frameTimings.startTime = start_time;
        }

        if (end_time) {
          this.frameTimings.endTime = end_time;
        }
      }
    } else if (updates) {
      if (updates.length === 0 || !updates.every(update => typeof update.timestamp === 'number')) {
        throw new Error('XVIZ updates did not contain a valid timestamp');
      }

      const min = Math.min(updates.map(update => update.timestamp));
      const max = Math.max(updates.map(update => update.timestamp));
      this.frameTimings.frames.set(index, [min, max, index, frameName(index)]);
    } else {
      // Missing updates & index is invalid call
      throw new Error('Cannot find timestamp');
    }
  }
  /* eslint-enable camelcase */


}
