// @streetscape.gl/monochrome/dist/es5/table/styled-components.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Expander = exports.TableCell = exports.TableRowComponent = exports.SortIcon = exports.HeaderCell = exports.HeaderContainer = exports.TableBody = exports.WrapperComponent = void 0;

var _styled = _interopRequireDefault(require("@emotion/styled"));

var _theme = require("../shared/theme");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// Table
var WrapperComponent = _styled.default.div(function (props) {
  return _objectSpread({}, props.theme.__reset__, {
    display: 'flex',
    flexDirection: 'column'
  }, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.WrapperComponent = WrapperComponent;

var TableBody = _styled.default.div(function (props) {
  return _objectSpread({
    flex: '1 1 auto'
  }, (0, _theme.evaluateStyle)(props.userStyle, props));
}); // TableHeader


exports.TableBody = TableBody;

var HeaderContainer = _styled.default.div(function (props) {
  return _objectSpread({
    display: 'table',
    flex: '0 0 auto'
  }, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.HeaderContainer = HeaderContainer;

var HeaderCell = _styled.default.div(function (props) {
  return _objectSpread({
    cursor: 'pointer',
    display: 'table-cell',
    position: 'relative',
    fontWeight: 'bold',
    background: props.isAscending || props.isDescending ? props.theme.controlColorActive : props.theme.controlColorPrimary,
    color: props.theme.textColorInvert,
    paddingTop: props.theme.spacingTiny,
    paddingBottom: props.theme.spacingTiny,
    paddingLeft: props.theme.spacingSmall,
    paddingRight: props.theme.spacingSmall,
    '&:hover': {
      background: props.theme.controlColorHovered
    }
  }, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.HeaderCell = HeaderCell;

var SortIcon = _styled.default.div(function (props) {
  return _objectSpread({
    position: 'absolute',
    right: props.theme.spacingTiny,
    bottom: props.theme.spacingTiny
  }, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.SortIcon = SortIcon;

var TableRowComponent = _styled.default.div(function (props) {
  return _objectSpread({
    position: 'relative',
    display: 'flex',
    alignItems: 'stretch',
    borderBottomStyle: 'solid',
    borderBottomWidth: 1,
    borderBottomColor: props.theme.controlColorSecondary
  }, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.TableRowComponent = TableRowComponent;

var TableCell = _styled.default.div(function (props) {
  return _objectSpread({
    flex: '0 0 auto',
    overflow: 'hidden',
    boxSizing: 'border-box',
    position: 'relative',
    paddingTop: props.theme.spacingTiny,
    paddingBottom: props.theme.spacingTiny,
    paddingLeft: props.theme.spacingSmall,
    paddingRight: props.theme.spacingSmall,
    borderLeftStyle: 'solid',
    borderLeftWidth: props.index === 0 ? 0 : 1,
    borderLeftColor: props.theme.controlColorSecondary
  }, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.TableCell = TableCell;

var Expander = _styled.default.div(function (props) {
  return _objectSpread({
    position: 'absolute',
    cursor: 'pointer',
    left: -props.theme.spacingSmall
  }, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.Expander = Expander;
