// @streetscape.gl/monochrome/dist/es5/table/tree-table-row.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = require("./styled-components");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * A stateless component that renders a data row in the TreeTable component
 */
var TreeTableRow =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(TreeTableRow, _PureComponent);

  function TreeTableRow() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, TreeTableRow);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(TreeTableRow)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_renderItem", function (_ref) {
      var id = _ref.id,
          index = _ref.index,
          _ref$depth = _ref.depth,
          depth = _ref$depth === void 0 ? 0 : _ref$depth,
          key = _ref.key,
          data = _ref.data,
          style = _ref.style;
      var _this$props = _this.props,
          indentSize = _this$props.indentSize,
          columns = _this$props.columns,
          theme = _this$props.theme,
          userStyle = _this$props.userStyle,
          renderCell = _this$props.renderCell,
          getIsExpanded = _this$props.getIsExpanded;
      var isExpanded = getIsExpanded(id);
      var hasChildren = data.children.length > 0;
      var indent = indentSize * (depth + 1);
      return _react.default.createElement("div", {
        key: key,
        style: style
      }, _react.default.createElement(_styledComponents.TableRowComponent, {
        theme: theme,
        index: index,
        userStyle: userStyle.row
      }, hasChildren && _react.default.createElement(_styledComponents.Expander, {
        key: "toggle",
        isExpanded: isExpanded,
        theme: theme,
        userStyle: userStyle.expander,
        style: {
          marginLeft: indent
        },
        onClick: function onClick() {
          return _this._toggleExpansion(id);
        }
      }, isExpanded ? userStyle.iconExpanded || '▾' : userStyle.iconCollapsed || '▸'), _react.default.createElement("div", {
        style: {
          flexShrink: 0,
          width: indent
        }
      }), data.data.map(function (colValue, colIndex) {
        var column = columns[colIndex];
        var width = colIndex === 0 ? column.width - indent : column.width;
        return _react.default.createElement(_styledComponents.TableCell, {
          key: colIndex,
          index: colIndex,
          style: {
            width: width
          },
          title: "".concat(column.name, ": ").concat(colValue),
          theme: theme,
          userStyle: userStyle.cell
        }, renderCell({
          value: colValue,
          column: column.srcObject,
          columnIndex: colIndex,
          row: data.srcObject,
          rowId: id
        }));
      })), hasChildren && isExpanded && data.children.map(function (row, rowIndex) {
        return _this._renderItem({
          depth: depth + 1,
          id: "".concat(id, ".").concat(rowIndex),
          index: rowIndex,
          key: rowIndex,
          data: row
        });
      }));
    });

    return _this;
  }

  _createClass(TreeTableRow, [{
    key: "_toggleExpansion",
    value: function _toggleExpansion(id) {
      this.props.toggleExpansion(id, this.props.id);
    }
  }, {
    key: "render",
    value: function render() {
      return this._renderItem(_objectSpread({}, this.props));
    }
  }]);

  return TreeTableRow;
}(_react.PureComponent);

exports.default = TreeTableRow;

_defineProperty(TreeTableRow, "propTypes", {
  id: _propTypes.default.string,
  data: _propTypes.default.shape({
    data: _propTypes.default.array,
    children: _propTypes.default.array
  }),
  style: _propTypes.default.object,
  indentSize: _propTypes.default.number,
  renderCell: _propTypes.default.func,
  columns: _propTypes.default.array,
  getIsExpanded: _propTypes.default.func,
  toggleExpansion: _propTypes.default.func
});
