// @xviz/parser/dist/esm/utils/worker-utils.js
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/* global Worker */
export function getTransferList(object) {
  var recursive = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
  var transfers = arguments.length > 2 ? arguments[2] : undefined;
  // Make sure that items in the transfer list is unique
  var transfersSet = transfers || new Set();

  if (!object) {// ignore
  } else if (object instanceof ArrayBuffer) {
    transfersSet.add(object);
  } else if (object.buffer && object.buffer instanceof ArrayBuffer) {
    // Typed array
    transfersSet.add(object.buffer);
  } else if (recursive && _typeof(object) === 'object') {
    for (var key in object) {
      // Avoid perf hit - only go one level deep
      getTransferList(object[key], false, transfersSet);
    }
  } // If transfers is defined, is internal recursive call
  // Otherwise it's called by the user


  return transfers === undefined ? Array.from(transfersSet) : null;
}
/**
 * A worker in the WorkerFarm
 */

var WorkerThread =
/*#__PURE__*/
function () {
  function WorkerThread(_ref) {
    var url = _ref.url,
        metadata = _ref.metadata;

    _classCallCheck(this, WorkerThread);

    this.worker = new Worker(url);
    this.isBusy = false;
    this.metadata = metadata;
  }

  _createClass(WorkerThread, [{
    key: "process",
    value: function process(data) {
      var _this = this;

      var worker = this.worker;
      return new Promise(function (resolve, reject) {
        worker.onmessage = function (e) {
          _this.isBusy = false; // console.log(e.data._size, `${Date.now() - e.data._sentAt}ms`);

          resolve(e.data);
        };

        worker.onerror = function (err) {
          _this.isBusy = false;
          reject(err);
        };

        _this.isBusy = true;
        worker.postMessage(data, getTransferList(data));
      });
    }
  }, {
    key: "terminate",
    value: function terminate() {
      this.worker.terminate();
      this.worker = null;
    }
  }]);

  return WorkerThread;
}();
/**
 * Process multiple data messages with a fleet of workers
 */


export var WorkerFarm =
/*#__PURE__*/
function () {
  /**
   * @param workerURL {function | string} - worker function
   * @param maxConcurrency {number} - max count of workers
   */
  function WorkerFarm(_ref2) {
    var workerURL = _ref2.workerURL,
        _ref2$maxConcurrency = _ref2.maxConcurrency,
        maxConcurrency = _ref2$maxConcurrency === void 0 ? 1 : _ref2$maxConcurrency,
        _ref2$debug = _ref2.debug,
        debug = _ref2$debug === void 0 ? function () {} : _ref2$debug,
        _ref2$initialMessage = _ref2.initialMessage,
        initialMessage = _ref2$initialMessage === void 0 ? null : _ref2$initialMessage;

    _classCallCheck(this, WorkerFarm);

    this.workerURL = workerURL;
    this.workers = [];
    this.queue = [];
    this.debug = debug;

    for (var i = 0; i < maxConcurrency; i++) {
      this.workers[i] = new WorkerThread({
        url: this.workerURL,
        metadata: {
          name: "".concat(i, "/").concat(maxConcurrency)
        }
      });
    }

    if (initialMessage) {
      this.broadcast(initialMessage);
    }
  }

  _createClass(WorkerFarm, [{
    key: "destroy",
    value: function destroy() {
      this.workers.forEach(function (worker) {
        return worker.terminate();
      });
    }
  }, {
    key: "getAvailableWorker",
    value: function getAvailableWorker() {
      return this.workers.find(function (worker) {
        return !worker.isBusy;
      });
    }
  }, {
    key: "broadcast",
    value: function broadcast(data) {
      var count = this.workers.length; // queue in reverse order as bias worker searching in getAvailableWorker()

      for (var i = count - 1; i >= 0; i--) {
        this.workers[i].worker.postMessage(data, getTransferList(data));
      }
    }
  }, {
    key: "next",
    value: function next() {
      var _this2 = this;

      var queue = this.queue;

      while (queue.length) {
        var worker = this.getAvailableWorker();

        if (!worker) {
          break;
        }

        var job = queue.shift();
        this.debug({
          message: 'processing',
          worker: worker.metadata.name,
          backlog: queue.length
        });
        worker.process(job.data).then(job.onResult).catch(job.onError).then(function () {
          return _this2.next();
        });
      }
    }
  }, {
    key: "process",
    value: function process(data, onResult, onError) {
      this.queue.push({
        data: data,
        onResult: onResult,
        onError: onError
      });
      this.next();
    }
  }]);

  return WorkerFarm;
}();
