// @loaders.gl/core/dist/esm/worker-utils/parse-with-worker.js
import { toArrayBuffer } from '../binary-utils/binary-utils';
var workerCache = new Map();
/* global Worker, Blob, URL */

function getWorker(workerSource) {
  var workerURL = workerCache.get(workerSource);

  if (!workerURL) {
    var blob = new Blob([workerSource], {
      type: 'application/javascript'
    });
    workerURL = URL.createObjectURL(blob);
    workerCache.set(workerSource, workerURL);
  }

  return new Worker(workerURL);
}

export default function parseWithWorker(workerSource, data, options) {
  var worker = getWorker(workerSource);
  options = removeNontransferableOptions(options);

  var parse = function parse(rawData, opts) {
    return new Promise(function (resolve, reject) {
      worker.onmessage = function (evt) {
        switch (evt.data.type) {
          case 'done':
            resolve(evt.data.result);
            worker.terminate();
            break;

          case 'error':
            reject(new Error(evt.data.message));
            break;

          default:
        }
      };

      var arraybuffer = toArrayBuffer(rawData);
      worker.postMessage({
        arraybuffer: arraybuffer,
        opts: opts
      }, [arraybuffer]);
    });
  };

  return data ? parse(data, options) : parse;
}

function removeNontransferableOptions(options) {
  options = Object.assign({}, options); // log object contains functions which cannot be transferred
  // TODO - decide how to handle logging on workers

  if (options.log !== null) {
    delete options.log;
  }

  return options;
}
