// @streetscape.gl/monochrome/dist/es6/form/input.js
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { CheckBox, Dropdown, RadioBox, Slider, TextBox, Toggle, Label } from '../shared';
import { Title, Heading, Separator } from './styled-components';
import styled from '@emotion/styled';
import { evaluateStyle } from '../shared/theme';
const InputContainer = styled.div(props => _objectSpread({
  display: 'flex',
  alignItems: 'start',
  justifyContent: 'space-between',
  width: '100%',
  boxSizing: 'border-box',
  paddingLeft: props.level * props.theme.spacingLarge,
  marginBottom: props.theme.spacingSmall,
  '>label': {
    marginTop: props.theme.spacingTiny,
    marginRight: props.theme.spacingSmall
  },
  '>label + div': {
    flexGrow: 1,
    maxWidth: 320
  }
}, evaluateStyle(props.userStyle, props)));
export default class Input extends PureComponent {
  constructor(_props) {
    var _this;

    _this = super(_props);

    _defineProperty(this, "_onChange", value => {
      this.props.onChange(this.props.name, value);
    });

    _defineProperty(this, "_renderTitle", (props, userStyle) => {
      return React.createElement(Title, _extends({}, props, {
        userStyle: userStyle
      }), props.title);
    });

    _defineProperty(this, "_renderHeading", (props, userStyle) => {
      return React.createElement(Heading, _extends({}, props, {
        userStyle: userStyle
      }), props.title);
    });

    _defineProperty(this, "_renderSeparator", (props, userStyle) => {
      return React.createElement(Separator, _extends({}, props, {
        userStyle: userStyle
      }));
    });

    _defineProperty(this, "_renderToggle", (props, style) => {
      const _this$props = this.props,
            label = _this$props.label,
            onTitle = _this$props.onTitle,
            offTitle = _this$props.offTitle,
            value = _this$props.value,
            className = _this$props.className;
      const labelText = (value ? onTitle : offTitle) || label;
      return React.createElement(Toggle, _extends({}, props, {
        style: style,
        className: className,
        label: this._renderLabel(labelText),
        onChange: this._onChange
      }));
    });

    _defineProperty(this, "_renderSlider", (props, style) => {
      return [this._renderLabel(), React.createElement(Slider, _extends({
        key: "slider"
      }, props, {
        style: style,
        onChange: this._onChange
      }))];
    });

    _defineProperty(this, "_renderDropdown", (props, style) => {
      return [this._renderLabel(), React.createElement(Dropdown, _extends({
        key: "dropdown"
      }, props, {
        style: style,
        onChange: this._onChange
      }))];
    });

    _defineProperty(this, "_renderRadio", (props, style) => {
      return [this._renderLabel(), React.createElement(RadioBox, _extends({
        key: "radio"
      }, props, {
        style: style,
        onChange: this._onChange
      }))];
    });

    _defineProperty(this, "_renderTextBox", (props, style) => {
      return [this._renderLabel(), React.createElement(TextBox, _extends({
        key: "textbox"
      }, props, {
        style: style,
        onChange: this._onChange
      }))];
    });

    _defineProperty(this, "_renderCheckbox", (props, style) => {
      return React.createElement(CheckBox, _extends({}, props, {
        label: this._renderLabel(),
        style: style,
        onChange: this._onChange
      }));
    });

    _defineProperty(this, "_renderCustom", () => {
      return this.props.render(this.props);
    });

    _defineProperty(this, "_renderLabel", function () {
      let label = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _this.props.label;
      const _this$props2 = _this.props,
            isEnabled = _this$props2.isEnabled,
            tooltip = _this$props2.tooltip,
            badge = _this$props2.badge,
            style = _this$props2.style;
      return label && React.createElement(Label, {
        key: "label",
        isEnabled: isEnabled,
        tooltip: tooltip,
        badge: badge,
        style: style.label
      }, label);
    });

    this.renders = {
      title: this._renderTitle,
      header: this._renderHeading,
      separator: this._renderSeparator,
      toggle: this._renderToggle,
      text: this._renderTextBox,
      range: this._renderSlider,
      select: this._renderDropdown,
      radio: this._renderRadio,
      checkbox: this._renderCheckbox,
      custom: this._renderCustom
    };
  }

  /* eslint-disable no-unused-vars */
  render() {
    const _this$props3 = this.props,
          style = _this$props3.style,
          type = _this$props3.type,
          onChange = _this$props3.onChange,
          label = _this$props3.label,
          tooltip = _this$props3.tooltip,
          badge = _this$props3.badge,
          otherProps = _objectWithoutProperties(_this$props3, ["style", "type", "onChange", "label", "tooltip", "badge"]);

    const render = this.renders[type];

    if (!render) {
      throw new Error(`Unknown setting type ${type}`);
    }

    const inputStyle = style[type];
    return React.createElement(InputContainer, _extends({}, otherProps, {
      userStyle: style.item
    }), render(otherProps, inputStyle));
  }

}

_defineProperty(Input, "propTypes", {
  onChange: PropTypes.func.isRequired,
  name: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  level: PropTypes.number.isRequired
});

_defineProperty(Input, "defaultProps", {
  className: ''
});
