// @streetscape.gl/core/dist/es5/components/declarative-ui/xviz-metric.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _monochrome = require("@streetscape.gl/monochrome");

var _constants = require("./constants");

var _connect = _interopRequireDefault(require("../connect"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var defaultFormatValue = function defaultFormatValue(x) {
  return Number.isFinite(x) ? x.toFixed(3) : String(x);
};

var XVIZMetricComponent =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(XVIZMetricComponent, _PureComponent);

  function XVIZMetricComponent() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, XVIZMetricComponent);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(XVIZMetricComponent)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onClick", function (x) {
      var _this$props = _this.props,
          onClick = _this$props.onClick,
          log = _this$props.log;

      if (onClick) {
        onClick(x);
      } else if (log) {
        log.seek(x);
      }
    });

    return _this;
  }

  _createClass(XVIZMetricComponent, [{
    key: "_extractDataProps",
    value: function _extractDataProps() {
      var _this$props2 = this.props,
          streams = _this$props2.streams,
          timeSeries = _this$props2.timeSeries;
      var data = {};
      var sampleSeries = null;
      streams.forEach(function (streamName, i) {
        var dataSeries = timeSeries[streamName];

        if (dataSeries) {
          sampleSeries = dataSeries;
          data[streamName] = dataSeries.valueSeries;
        }
      });
      return sampleSeries ? {
        getX: sampleSeries.getX,
        getY: sampleSeries.getY,
        unit: sampleSeries.unit,
        data: data
      } : {
        isLoading: true
      };
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props3 = this.props,
          title = _this$props3.title,
          description = _this$props3.description,
          startTime = _this$props3.startTime,
          endTime = _this$props3.endTime,
          currentTime = _this$props3.currentTime,
          width = _this$props3.width,
          height = _this$props3.height,
          style = _this$props3.style,
          xTicks = _this$props3.xTicks,
          yTicks = _this$props3.yTicks,
          formatXTick = _this$props3.formatXTick,
          formatYTick = _this$props3.formatYTick,
          formatValue = _this$props3.formatValue,
          horizontalGridLines = _this$props3.horizontalGridLines,
          verticalGridLines = _this$props3.verticalGridLines,
          getColor = _this$props3.getColor;
      var isLoading = currentTime === null;
      var timeDomain = Number.isFinite(startTime) ? [startTime, endTime] : null;
      return _react.default.createElement(_monochrome.MetricCard, {
        title: title,
        description: description,
        isLoading: isLoading,
        style: style
      }, !isLoading && _react.default.createElement(_monochrome.MetricChart, _extends({}, this._extractDataProps(), {
        getColor: getColor,
        highlightX: currentTime,
        width: width,
        height: height,
        style: style,
        xTicks: xTicks,
        yTicks: yTicks,
        formatXTick: formatXTick,
        formatYTick: formatYTick,
        formatValue: formatValue,
        xDomain: timeDomain,
        onClick: this._onClick,
        horizontalGridLines: horizontalGridLines,
        verticalGridLines: verticalGridLines
      })));
    }
  }]);

  return XVIZMetricComponent;
}(_react.PureComponent);

_defineProperty(XVIZMetricComponent, "propTypes", {
  // User configuration
  style: _propTypes.default.object,
  width: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  height: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number]),
  getColor: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.string, _propTypes.default.object]),
  xTicks: _propTypes.default.number,
  yTicks: _propTypes.default.number,
  formatXTick: _propTypes.default.func,
  formatYTick: _propTypes.default.func,
  formatValue: _propTypes.default.func,
  horizontalGridLines: _propTypes.default.number,
  verticalGridLines: _propTypes.default.number,
  onClick: _propTypes.default.func,
  // From declarative UI metric component
  streams: _propTypes.default.arrayOf(_propTypes.default.string),
  title: _propTypes.default.string,
  description: _propTypes.default.string,
  // From connected log
  currentTime: _propTypes.default.number,
  timeSeries: _propTypes.default.objectOf(_propTypes.default.shape({
    data: _propTypes.default.array,
    getX: _propTypes.default.func,
    getY: _propTypes.default.func,
    unit: _propTypes.default.string
  })),
  startTime: _propTypes.default.number,
  endTime: _propTypes.default.number
});

_defineProperty(XVIZMetricComponent, "defaultProps", {
  timeSeries: {},
  width: '100%',
  height: 160,
  style: {
    margin: {
      left: 45,
      right: 10,
      top: 10,
      bottom: 20
    }
  },
  xTicks: 0,
  yTicks: 3,
  formatValue: defaultFormatValue,
  horizontalGridLines: 3,
  verticalGridLines: 0,
  getColor: _constants.DEFAULT_COLOR_SERIES
});

var getLogState = function getLogState(log) {
  return {
    currentTime: log.getCurrentTime(),
    timeSeries: log.getTimeSeries(),
    startTime: log.getBufferStart(),
    endTime: log.getBufferEnd()
  };
};

var _default = (0, _connect.default)({
  getLogState: getLogState,
  Component: XVIZMetricComponent
});

exports.default = _default;
