// @streetscape.gl/core/dist/es5/loaders/xviz-websocket-loader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _assert = _interopRequireDefault(require("assert"));

var _parser = require("@xviz/parser");

var _promiseRetry = _interopRequireDefault(require("promise-retry"));

var _xvizLoaderInterface = _interopRequireDefault(require("./xviz-loader-interface"));

var _xvizControllerV = _interopRequireDefault(require("./xviz-controller-v2"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _get(target, property, receiver) { if (typeof Reflect !== "undefined" && Reflect.get) { _get = Reflect.get; } else { _get = function _get(target, property, receiver) { var base = _superPropBase(target, property); if (!base) return; var desc = Object.getOwnPropertyDescriptor(base, property); if (desc.get) { return desc.get.call(receiver); } return desc.value; }; } return _get(target, property, receiver || target); }

function _superPropBase(object, property) { while (!Object.prototype.hasOwnProperty.call(object, property)) { object = _getPrototypeOf(object); if (object === null) break; } return object; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * Connect to XVIZ 2 websocket manage storage of XVIZ data into a XVIZStreamBuffer
 *
 * This class is a Websocket base class and is expected to be subclassed with
 * the following methods overridden:
 *
 * - _onOpen()
 * - _onXVIZTimeslice()
 */
var XVIZWebsocketLoader =
/*#__PURE__*/
function (_XVIZLoaderInterface) {
  _inherits(XVIZWebsocketLoader, _XVIZLoaderInterface);

  /**
   * constructor
   * @params serverConfig {object}
   *   - serverConfig.serverUrl {string}
   *   - serverConfig.defaultLogLength {number, optional} - default 30
   *   - serverConfig.queryParams {object, optional}
   *   - serverConfig.retryAttempts {number, optional} - default 3
   * @params worker {string|function, optional}
   * @params maxConcurrency {number, optional} - default 3
   * @params logGuid {string}
   * @params logProfile {string, optional}
   * @params duration {number, optional}
   * @params timestamp {number, optional}
   * @params bufferLength {number, optional}
   */
  function XVIZWebsocketLoader() {
    var _this;

    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    _classCallCheck(this, XVIZWebsocketLoader);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(XVIZWebsocketLoader).call(this, options));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onWSOpen", function () {
      // Request data if we are restarting, otherwise wait for metadata
      // TODO - protocol negotiation
      _this.xvizHandler = new _xvizControllerV.default(_this.socket);

      _this._debug('socket_open', _this.requestParams);

      _this._onOpen();
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onWSMessage", function (message) {
      switch (message.type) {
        case _parser.LOG_STREAM_MESSAGE.METADATA:
          if (_this.get('metadata')) {
            // already has metadata
            return;
          }

          _this.set('logSynchronizer', new _parser.StreamSynchronizer(_this.streamBuffer));

          _this._setMetadata(message);

          _this.emit('ready', message);

          break;

        case _parser.LOG_STREAM_MESSAGE.TIMESLICE:
          var oldVersion = _this.streamBuffer.valueOf();

          _this.streamBuffer.insert(message);

          var bufferUpdated = _this.streamBuffer.valueOf() !== oldVersion;

          if (bufferUpdated) {
            _this.set('streams', _this.streamBuffer.getStreams());
          }

          _this._onXVIZTimeslice(message, bufferUpdated);

          _this.emit('update', message);

          break;

        case _parser.LOG_STREAM_MESSAGE.DONE:
          _this.emit('finish', message);

          break;

        default:
          _this.emit('error', message);

      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onWSError", function (error) {
      _this.emit('error', error);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onWSClose", function (event) {
      // Only called on connection closure, which would be an error case
      _this._debug('socket_closed', event);
    });

    _this.socket = null;
    _this.retrySettings = {
      retries: 3,
      minTimeout: 500,
      randomize: true
    };
    _this.streamBuffer = new _parser.XVIZStreamBuffer(); // Handler object for the websocket events
    // Note: needs to be last due to member dependencies

    _this.WebSocketClass = options.WebSocketClass || WebSocket;
    return _this;
  }

  _createClass(XVIZWebsocketLoader, [{
    key: "isOpen",
    value: function isOpen() {
      return this.socket; // && this.socket.readyState === WEB_SOCKET_OPEN_STATE;
    }
  }, {
    key: "seek",
    value: function seek(timestamp) {
      _get(_getPrototypeOf(XVIZWebsocketLoader.prototype), "seek", this).call(this, timestamp);
    }
    /**
     * Open an XVIZ socket connection with automatic retry
     *
     * @returns {Promise} WebSocket connection
     */

  }, {
    key: "connect",
    value: function connect() {
      var _this2 = this;

      (0, _assert.default)(this.socket === null, 'Socket Manager still connected');

      this._debug('stream_start');

      var url = this.requestParams.url; // Wrap retry logic around connection

      return (0, _promiseRetry.default)(function (retry) {
        return new Promise(function (resolve, reject) {
          try {
            var ws = new _this2.WebSocketClass(url);
            ws.binaryType = 'arraybuffer';

            ws.onmessage = function (message) {
              var hasMetadata = Boolean(_this2.getMetadata());
              return (0, _parser.parseStreamMessage)({
                message: message.data,
                onResult: _this2._onWSMessage,
                onError: _this2._onWSError,
                debug: _this2._debug.bind('parse_message'),
                worker: hasMetadata && _this2.options.worker,
                maxConcurrency: _this2.options.maxConcurrency
              });
            };

            ws.onerror = _this2._onWSError;

            ws.onclose = function (event) {
              _this2._onWSClose(event);

              reject(event);
            }; // On success, resolve the promise with the now ready socket


            ws.onopen = function () {
              _this2.socket = ws;

              _this2._onWSOpen();

              resolve(ws);
            };
          } catch (err) {
            reject(err);
          }
        }).catch(function (event) {
          _this2._onWSError(event);

          var isAbnormalClosure = event.code > 1000 && event.code !== 1005; // Retry if abnormal or connection never established

          if (isAbnormalClosure || !_this2.socket) {
            retry();
          }
        });
      }, this.retrySettings).catch(this._onWSError);
    }
  }, {
    key: "close",
    value: function close() {
      if (this.socket) {
        this.socket.close();
        this.socket = null;
      }
    } // Subclasses *MUST* implement these methods

  }, {
    key: "_onOpen",
    value: function _onOpen() {
      throw new Error('_onOpen() method must be overridden');
    }
    /**
     * Subclass hook for xviz message
     *
     * The messages will be inserted into the 'streamBuffer'
     * prior to this hook being called.
     *
     * @params message {Object} Parsed XVIZ message
     * @params bufferUpdated {Boolean} True if streamBuffer has changed
     */

  }, {
    key: "_onXVIZTimeslice",
    value: function _onXVIZTimeslice(message, bufferUpdated) {} // PRIVATE Methods
    // Notifications and metric reporting

  }]);

  return XVIZWebsocketLoader;
}(_xvizLoaderInterface.default);

exports.default = XVIZWebsocketLoader;
