// @streetscape.gl/core/dist/es5/cars/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _sedan2 = _interopRequireDefault(require("./sedan"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Copyright (c) 2019 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
function createCar(makeMesh, opts) {
  var length = opts.length,
      width = opts.width,
      height = opts.height,
      _opts$color = opts.color,
      color = _opts$color === void 0 ? [128, 128, 128] : _opts$color,
      _opts$origin = opts.origin,
      origin = _opts$origin === void 0 ? [0, 0, 0] : _opts$origin;

  if (!Number.isFinite(length) || !Number.isFinite(width) || !Number.isFinite(height)) {
    throw new Error('invalid dimensions');
  }

  var mesh = makeMesh(length, width, height);
  return {
    color: color,
    origin: origin,
    mesh: mesh,
    scale: [1, 1, 1]
  };
}

var _default = {
  sedan: function sedan(opts) {
    return createCar(_sedan2.default, opts);
  }
};
exports.default = _default;
