// @streetscape.gl/monochrome/dist/es6/shared/draggable/index.js
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { evaluateStyle } from '../theme';
const ContainerComponent = styled.div(props => _objectSpread({
  margin: -props.tolerance,
  padding: props.tolerance,
  cursor: props.isActive ? 'grabbing' : props.isEnabled ? 'grab' : 'inherit'
}, evaluateStyle(props.userStyle, props)));
const BACKDROP_STYLES = {
  position: 'fixed',
  zIndex: 999,
  top: 0,
  left: 0,
  width: '100%',
  height: '100%'
};

function noop() {}
/**
 * @class
 */


export default class Draggable extends PureComponent {
  constructor(props) {
    var _this;

    _this = super(props);

    _defineProperty(this, "_getEventData", function (evt) {
      let offset = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _this.state.offset;
      const _this$state = _this.state,
            dragStartPos = _this$state.dragStartPos,
            hasDragged = _this$state.hasDragged;
      const result = {
        srcEvent: evt,
        x: evt.clientX,
        y: evt.clientY,
        offsetX: evt.clientX - offset.left,
        offsetY: evt.clientY - offset.top,
        hasDragged
      };

      if (dragStartPos) {
        result.deltaX = result.x - dragStartPos.x;
        result.deltaY = result.y - dragStartPos.y;
      } else {
        result.deltaX = 0;
        result.deltaY = 0;
      }

      return result;
    });

    _defineProperty(this, "_onMouseDown", evt => {
      if (!this.props.isEnabled) {
        return;
      }

      evt.stopPropagation();

      const offset = this._element.getBoundingClientRect();

      const eventData = this._getEventData(evt, offset);

      this.setState({
        isMouseDown: true,
        hasDragged: false,
        offset,
        dragStartPos: {
          x: eventData.x,
          y: eventData.y
        }
      });
      this.props.onDragStart(eventData);
    });

    _defineProperty(this, "_onMouseMove", evt => {
      if (!this.props.isEnabled) {
        return;
      }

      evt.stopPropagation();

      if (this.state.isMouseDown) {
        const eventData = this._getEventData(evt);

        const deltaX = eventData.deltaX,
              deltaY = eventData.deltaY;

        if (!this.state.hasDragged) {
          if (deltaX || deltaY) {
            this.setState({
              hasDragged: true
            });
          } else {
            return;
          }
        }

        this.props.onDrag(eventData);
      }
    });

    _defineProperty(this, "_onMouseUp", evt => {
      if (this.state.isMouseDown) {
        this.setState({
          isMouseDown: false,
          dragStartPos: null
        });
        this.props.onDragEnd(this._getEventData(evt));
      }
    });

    this.state = {
      isMouseDown: false,
      dragStartPos: null,
      hasDragged: false
    };
  }

  render() {
    const _this$props = this.props,
          style = _this$props.style,
          isEnabled = _this$props.isEnabled,
          className = _this$props.className,
          tolerance = _this$props.tolerance;
    const isMouseDown = this.state.isMouseDown;
    return React.createElement(ContainerComponent, {
      className: className,
      ref: _ref => {
        this._element = _ref;
      },
      tolerance: tolerance,
      isEnabled: isEnabled,
      isActive: isMouseDown,
      userStyle: style,
      onMouseDown: this._onMouseDown,
      onMouseMove: this._onMouseMove,
      onMouseLeave: this._onMouseUp,
      onMouseUp: this._onMouseUp
    }, isMouseDown && React.createElement("div", {
      style: BACKDROP_STYLES
    }), this.props.children);
  }

}

_defineProperty(Draggable, "propTypes", {
  // container
  className: PropTypes.string,
  // config
  style: PropTypes.object,
  tolerance: PropTypes.number,
  isEnabled: PropTypes.bool,
  // callbacks
  onDragStart: PropTypes.func,
  onDrag: PropTypes.func,
  onDragEnd: PropTypes.func
});

_defineProperty(Draggable, "defaultProps", {
  className: '',
  isEnabled: true,
  tolerance: 0,
  onDragStart: noop,
  onDrag: noop,
  onDragEnd: noop
});
