// @streetscape.gl/monochrome/dist/es6/playback-control/index.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import AutoSizer from '../shared/autosizer';
import Slider from '../shared/slider';
import { getTimelineTicks, formatTimeCode } from './utils';
import { scaleLinear } from 'd3-scale';
import { withTheme } from '../shared/theme';
import { WrapperComponent, ControlsContainer, PlayPauseButton, Timestamp, TicksContainer, Tick, TickLabel, MarkersContainer, MarkerComponent, BufferComponent } from './styled-components.js';
import { PlayIcon, PauseIcon } from '../shared/icons';
const DEFAULT_PADDING = 24;
const COMPACT_CONTAINER_STYLE = {
  display: 'flex',
  alignItems: 'flex-end'
};

function noop() {}

function normalizePadding(padding) {
  padding = padding || 0;

  if (Number.isFinite(padding)) {
    return {
      right: padding,
      left: padding
    };
  }

  return Object.assign({
    right: 0,
    left: 0
  }, padding);
}
/*
 * @class
 */


class PlaybackControl extends PureComponent {
  constructor(props) {
    super(props);

    _defineProperty(this, "_play", () => {
      this.props.onPlay();
    });

    _defineProperty(this, "_pause", () => {
      this.props.onPause();
    });

    _defineProperty(this, "_seek", newTime => {
      const currentTime = this.props.currentTime;

      if (newTime !== currentTime) {
        this.props.onSeek(newTime);
      }
    });

    _defineProperty(this, "_onResize", (_ref) => {
      let width = _ref.width;
      let _this$props$style$pad = this.props.style.padding,
          padding = _this$props$style$pad === void 0 ? 24 : _this$props$style$pad;
      padding = normalizePadding(padding);
      this.scale.range([0, width - padding.left - padding.right]); // Trigger rerender

      this.setState({
        width
      });
    });

    this.scale = scaleLinear().domain([props.startTime, props.endTime]);
  }

  UNSAFE_componentWillReceiveProps(newProps) {
    const props = this.props;

    if (newProps.step !== props.step || newProps.startTime !== props.startTime || newProps.endTime !== props.endTime) {
      // Video source has changed
      // Kill any running animation to avoid callbacks in incorrect time range
      this._pause(); // Update currentTime to make sure it is the start of the new range


      this._seek(newProps.startTime);

      this.scale.domain([newProps.startTime, newProps.endTime]);
    }
  }

  componentWillUnmount() {
    this._pause();
  }

  _renderMarker(marker, i, Component, styleProps, userStyle) {
    const scale = this.scale;
    const _marker$startTime = marker.startTime,
          startTime = _marker$startTime === void 0 ? marker.time : _marker$startTime,
          _marker$endTime = marker.endTime,
          endTime = _marker$endTime === void 0 ? marker.time : _marker$endTime,
          style = marker.style,
          content = marker.content;
    const x0 = scale(startTime);
    const x1 = scale(endTime);

    const markerStyle = _objectSpread({}, style, {
      position: 'absolute',
      left: x0,
      width: x1 - x0
    });

    return React.createElement(Component, _extends({
      key: i
    }, styleProps, {
      style: markerStyle,
      userStyle: userStyle
    }), content);
  }

  _renderTimeline(styleProps) {
    const _this$props = this.props,
          style = _this$props.style,
          tickSpacing = _this$props.tickSpacing,
          formatTick = _this$props.formatTick,
          markers = _this$props.markers;
    const scale = this.scale;
    const ticks = getTimelineTicks(scale, tickSpacing, formatTick);
    return React.createElement("div", null, React.createElement(TicksContainer, _extends({}, styleProps, {
      userStyle: style.ticks
    }), ticks.map((t, i) => {
      const tickStyle = {
        position: 'absolute',
        left: t.x
      };
      return React.createElement(Tick, _extends({
        key: i
      }, styleProps, {
        userStyle: style.tick,
        style: tickStyle
      }), React.createElement(TickLabel, _extends({}, styleProps, {
        userStyle: style.tickLabel
      }), t.label));
    })), markers && React.createElement(MarkersContainer, _extends({}, styleProps, {
      userStyle: style.markers
    }), markers.map((marker, i) => this._renderMarker(marker, i, MarkerComponent, styleProps, style.marker))));
  }

  _renderSlider(styleProps) {
    const _this$props2 = this.props,
          style = _this$props2.style,
          currentTime = _this$props2.currentTime,
          startTime = _this$props2.startTime,
          endTime = _this$props2.endTime,
          step = _this$props2.step,
          _this$props2$bufferRa = _this$props2.bufferRange,
          bufferRange = _this$props2$bufferRa === void 0 ? [] : _this$props2$bufferRa;
    const buffers = Array.isArray(bufferRange) ? bufferRange : [bufferRange];
    return React.createElement(Slider, {
      style: style.slider,
      value: currentTime,
      onChange: this._seek,
      knobSize: 18,
      step: step,
      min: startTime,
      max: endTime
    }, buffers.map((range, i) => this._renderMarker(range, i, BufferComponent, styleProps, style.buffer)));
  }

  _renderPlayPauseButton(styleProps) {
    const _this$props3 = this.props,
          isPlaying = _this$props3.isPlaying,
          style = _this$props3.style;
    return React.createElement(PlayPauseButton, _extends({}, styleProps, {
      userStyle: style.playPauseButton,
      onClick: isPlaying ? this._pause : this._play
    }), isPlaying ? style.iconPause || React.createElement(PauseIcon, null) : style.iconPlay || React.createElement(PlayIcon, null));
  }

  _renderTimestamp(styleProps) {
    const _this$props4 = this.props,
          style = _this$props4.style,
          currentTime = _this$props4.currentTime,
          formatTimestamp = _this$props4.formatTimestamp;
    return React.createElement(Timestamp, _extends({}, styleProps, {
      userStyle: style.timestamp
    }), formatTimestamp(currentTime));
  }

  render() {
    const _this$props5 = this.props,
          theme = _this$props5.theme,
          compact = _this$props5.compact,
          width = _this$props5.width,
          style = _this$props5.style,
          className = _this$props5.className,
          isPlaying = _this$props5.isPlaying;
    let _style$padding = style.padding,
        padding = _style$padding === void 0 ? DEFAULT_PADDING : _style$padding;
    padding = normalizePadding(padding);
    const styleProps = {
      theme,
      compact,
      isPlaying
    };
    const wrapperStyle = {
      width
    };

    if (compact) {
      const sliderStyle = {
        flexGrow: 1,
        paddingLeft: padding.left,
        paddingRight: padding.right
      };
      return React.createElement(WrapperComponent, _extends({
        className: className
      }, styleProps, {
        userStyle: style.wrapper,
        style: wrapperStyle
      }), React.createElement("div", {
        style: COMPACT_CONTAINER_STYLE
      }, this._renderPlayPauseButton(styleProps), React.createElement("div", {
        style: sliderStyle
      }, React.createElement(AutoSizer, {
        disableHeight: true,
        onResize: this._onResize
      }), this._renderTimeline(styleProps), this._renderSlider(styleProps)), this._renderTimestamp(styleProps)), React.createElement(ControlsContainer, _extends({}, styleProps, {
        userStyle: style.controls
      }), this.props.children));
    }

    Object.assign(wrapperStyle, {
      paddingLeft: padding.left,
      paddingRight: padding.right
    });
    return React.createElement(WrapperComponent, _extends({
      className: className
    }, styleProps, {
      userStyle: style.wrapper,
      style: wrapperStyle
    }), React.createElement(AutoSizer, {
      disableHeight: true,
      onResize: this._onResize
    }), this._renderTimeline(styleProps), this._renderSlider(styleProps), React.createElement(ControlsContainer, _extends({}, styleProps, {
      userStyle: style.controls
    }), this._renderPlayPauseButton(styleProps), this._renderTimestamp(styleProps), this.props.children));
  }

}

_defineProperty(PlaybackControl, "propTypes", {
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  style: PropTypes.object,
  compact: PropTypes.bool,
  currentTime: PropTypes.number.isRequired,
  startTime: PropTypes.number,
  endTime: PropTypes.number.isRequired,
  isPlaying: PropTypes.bool.isRequired,
  className: PropTypes.string,
  // config
  step: PropTypes.number,
  tickSpacing: PropTypes.number,
  bufferRange: PropTypes.arrayOf(PropTypes.object),
  markers: PropTypes.arrayOf(PropTypes.object),
  // callbacks
  formatTick: PropTypes.func,
  formatTimestamp: PropTypes.func,
  onPlay: PropTypes.func,
  onPause: PropTypes.func,
  onSeek: PropTypes.func
});

_defineProperty(PlaybackControl, "defaultProps", {
  style: {},
  compact: false,
  className: '',
  startTime: 0,
  step: 0.1,
  tickSpacing: 100,
  formatTick: x => formatTimeCode(x, '{mm}:{ss}'),
  formatTimestamp: x => formatTimeCode(x, '{mm}:{ss}.{S}'),
  onPlay: noop,
  onPause: noop,
  onSeek: noop
});

const ThemedPlaybackControl = withTheme(PlaybackControl);
ThemedPlaybackControl.formatTimeCode = formatTimeCode;
export default ThemedPlaybackControl;
