// @streetscape.gl/monochrome/dist/es6/playback-control/styled-components.js
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import styled from '@emotion/styled';
import { evaluateStyle } from '../shared/theme';
export const WrapperComponent = styled.div(props => {
  const result = _objectSpread({}, props.theme.__reset__, {
    boxSizing: 'border-box'
  });

  if (props.isPlaying) {
    result.div = {
      transitionDuration: '0s !important'
    };
  }

  return Object.assign(result, evaluateStyle(props.userStyle, props));
});
export const ControlsContainer = styled.div(props => _objectSpread({
  display: 'flex',
  alignItems: 'center',
  marginTop: props.theme.spacingTiny
}, evaluateStyle(props.userStyle, props)));
export const PlayPauseButton = styled.div(props => _objectSpread({
  width: 16,
  height: 16,
  marginLeft: props.compact ? 0 : -8,
  marginRight: props.theme.spacingSmall,
  cursor: 'pointer',
  color: props.theme.controlColorPrimary,
  '&:hover': {
    color: props.theme.controlColorHovered
  },
  path: {
    fill: 'currentColor'
  }
}, evaluateStyle(props.userStyle, props)));
export const Timestamp = styled.div(props => _objectSpread({}, evaluateStyle(props.userStyle, props)));
export const TicksContainer = styled.div(props => _objectSpread({
  position: 'relative',
  height: 20
}, evaluateStyle(props.userStyle, props)));
export const Tick = styled.div(props => _objectSpread({
  height: 4,
  bottom: 0,
  borderLeftStyle: 'solid',
  borderLeftWidth: 1,
  borderLeftColor: props.theme.controlColorSecondary
}, evaluateStyle(props.userStyle, props)));
export const TickLabel = styled.div(props => _objectSpread({
  transform: 'translate(-50%, -18px)'
}, evaluateStyle(props.userStyle, props)));
export const MarkersContainer = styled.div(props => _objectSpread({
  position: 'relative',
  height: 3
}, evaluateStyle(props.userStyle, props)));
export const MarkerComponent = styled.div(props => _objectSpread({
  height: '100%'
}, evaluateStyle(props.userStyle, props)));
export const BufferComponent = styled.div(props => _objectSpread({
  height: '100%',
  background: props.theme.controlColorHovered
}, evaluateStyle(props.userStyle, props)));
