// @xviz/builder/dist/es5/builders/declarative-ui/xviz-ui-builder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _xvizPanelBuilder = _interopRequireDefault(require("./xviz-panel-builder"));

var _xvizContainerBuilder = _interopRequireDefault(require("./xviz-container-builder"));

var _xvizMetricBuilder = _interopRequireDefault(require("./xviz-metric-builder"));

var _xvizPlotBuilder = _interopRequireDefault(require("./xviz-plot-builder"));

var _xvizSelectBuilder = _interopRequireDefault(require("./xviz-select-builder"));

var _xvizTableBuilder = _interopRequireDefault(require("./xviz-table-builder"));

var _xvizTreeTableBuilder = _interopRequireDefault(require("./xviz-tree-table-builder"));

var _xvizVideoBuilder = _interopRequireDefault(require("./xviz-video-builder"));

var _utils = require("./utils");

var _constants = require("./constants");

var _UI_BUILDER_MAP;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var defaultValidateWarn = console.warn;
var defaultValidateError = console.error;
var UI_BUILDER_MAP = (_UI_BUILDER_MAP = {}, _defineProperty(_UI_BUILDER_MAP, _constants.UI_TYPES.PANEL, _xvizPanelBuilder.default), _defineProperty(_UI_BUILDER_MAP, _constants.UI_TYPES.CONTAINER, _xvizContainerBuilder.default), _defineProperty(_UI_BUILDER_MAP, _constants.UI_TYPES.METRIC, _xvizMetricBuilder.default), _defineProperty(_UI_BUILDER_MAP, _constants.UI_TYPES.PLOT, _xvizPlotBuilder.default), _defineProperty(_UI_BUILDER_MAP, _constants.UI_TYPES.SELECT, _xvizSelectBuilder.default), _defineProperty(_UI_BUILDER_MAP, _constants.UI_TYPES.TABLE, _xvizTableBuilder.default), _defineProperty(_UI_BUILDER_MAP, _constants.UI_TYPES.TREETABLE, _xvizTreeTableBuilder.default), _defineProperty(_UI_BUILDER_MAP, _constants.UI_TYPES.VIDEO, _xvizVideoBuilder.default), _UI_BUILDER_MAP);

var XVIZUIBuilder =
/*#__PURE__*/
function () {
  function XVIZUIBuilder() {
    var _this = this;

    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    _classCallCheck(this, XVIZUIBuilder);

    this._validateWarn = options.validateWarn || defaultValidateWarn;
    this._validateError = options.validateError || defaultValidateError;
    this._children = [];
    Object.values(_constants.UI_TYPES).map(function (type) {
      // add UI builders, e.g.
      // type `panel`
      //  - this.panel = (props) => this._setChild('panel', props);
      var camelType = (0, _utils.snakeToCamel)(type);

      _this[camelType] = function (props) {
        return _this._createUIBuilder(type, props);
      };
    });
  }

  _createClass(XVIZUIBuilder, [{
    key: "getUI",
    value: function getUI() {
      return this._children.reduce(function (ui, child) {
        var childConfig = child.getUI();
        ui[childConfig.name] = childConfig;
        return ui;
      }, {});
    }
  }, {
    key: "child",
    value: function child(_child) {
      if (!(_child instanceof _xvizPanelBuilder.default)) {
        this._validateError('Top level UI element should be `Panel`');
      }

      this._children.push(_child);

      return _child;
    }
  }, {
    key: "_createUIBuilder",
    value: function _createUIBuilder(type, props) {
      return new UI_BUILDER_MAP[type](_objectSpread({}, props, {
        validateWarn: this._validateWarn,
        validateError: this._validateError,
        root: this
      }));
    }
  }]);

  return XVIZUIBuilder;
}();

exports.default = XVIZUIBuilder;
