// @streetscape.gl/core/dist/es5/components/log-viewer/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LIGHT_SETTINGS = exports.CAR_DATA = exports.DEFAULT_ORIGIN = exports.DEFAULT_CAR = void 0;

var _luma = require("luma.gl");

// Copyright (c) 2019 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
var DEFAULT_CAR = {
  mesh: new _luma.CubeGeometry(),
  origin: [0, 0, 0.7],
  color: [128, 128, 128],
  scale: [2, 1, 0.7]
};
exports.DEFAULT_CAR = DEFAULT_CAR;
var DEFAULT_ORIGIN = [0, 0, 0];
exports.DEFAULT_ORIGIN = DEFAULT_ORIGIN;
var CAR_DATA = [[0, 0, 0]];
exports.CAR_DATA = CAR_DATA;
var LIGHT_SETTINGS = {
  lightsPosition: [0, 0, 5000, -1000, 400, 1000],
  ambientRatio: 0.5,
  diffuseRatio: 0.2,
  specularRatio: 0.4,
  lightsStrength: [0.68, 0.0, 1.2, 0.0],
  numberOfLights: 2
};
exports.LIGHT_SETTINGS = LIGHT_SETTINGS;
