// @streetscape.gl/monochrome/dist/es5/form/styled-components.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Separator = exports.Heading = exports.Title = exports.Expander = exports.Container = void 0;

var _styled = _interopRequireDefault(require("@emotion/styled"));

var _theme = require("../shared/theme");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Container = _styled.default.div(function (props) {
  return _objectSpread({}, props.theme.__reset__, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.Container = Container;

var Expander = _styled.default.div(function (props) {
  return _objectSpread({
    position: 'absolute',
    cursor: 'pointer',
    left: -20,
    top: 4,
    width: 16,
    height: 16,
    color: props.theme.controlColorPrimary,
    '&:hover': {
      color: props.theme.controlColorHovered
    },
    path: {
      fill: 'currentColor'
    }
  }, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.Expander = Expander;

var Title = _styled.default.div(function (props) {
  return _objectSpread({
    marginTop: props.theme.spacingNormal,
    marginBottom: props.theme.spacingNormal,
    fontSize: props.theme.fontSize * 2,
    fontWeight: 200
  }, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.Title = Title;

var Heading = _styled.default.div(function (props) {
  return _objectSpread({
    marginTop: props.theme.spacingNormal,
    marginBottom: props.theme.spacingNormal,
    fontSize: props.theme.fontSize * 1.2,
    fontWeight: 600
  }, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.Heading = Heading;

var Separator = _styled.default.div(function (props) {
  return _objectSpread({
    width: '100%',
    height: 1,
    background: props.theme.controlColorSecondary
  }, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.Separator = Separator;
