// @streetscape.gl/monochrome/dist/esm/table/table-header.js
function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import AutoSizer from '../shared/autosizer';
import { HeaderContainer, HeaderCell, SortIcon } from './styled-components';
var SORT = {
  NONE: 0,
  ASCEND: 1,
  DESCEND: 2
};

var TableHeader =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(TableHeader, _PureComponent);

  function TableHeader(props) {
    var _this;

    _classCallCheck(this, TableHeader);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(TableHeader).call(this, props));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_onResize", function () {
      var columns = _this.state.columns;
      columns.map(function (col, colIndex) {
        var ref = _this._cells[colIndex];

        if (ref) {
          col.width = ref.clientWidth;
        }
      });

      _this.props.onResize(columns);
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_sortColumn", function (index) {
      var columns = _this.state.columns;
      var sortType = columns[index].sort === SORT.ASCEND ? SORT.DESCEND : SORT.ASCEND;
      columns.forEach(function (col, colIndex) {
        col.sort = colIndex === index ? sortType : SORT.NONE;
      });
      var multiplier = sortType === SORT.ASCEND ? 1 : -1;

      var sortFunc = function sortFunc(row1, row2) {
        return row1.data[index] <= row2.data[index] ? -multiplier : multiplier;
      };

      _this.props.onSort(sortFunc); // Trigger rerender


      _this.forceUpdate();
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_renderColumn", function (column, colIndex) {
      var _this$props = _this.props,
          renderHeader = _this$props.renderHeader,
          theme = _this$props.theme,
          userStyle = _this$props.userStyle;
      var styleProps = {
        theme: theme,
        isAscending: column.sort === SORT.ASCEND,
        isDescending: column.sort === SORT.DESCEND
      };
      var icon = null;

      if (styleProps.isAscending) {
        icon = userStyle.iconAscending || '↑';
      } else if (styleProps.isDescending) {
        icon = userStyle.iconDescending || '↓';
      }

      return React.createElement(HeaderCell, _extends({}, styleProps, {
        userStyle: userStyle.headerCell,
        style: {
          width: column.defaultWidth
        },
        key: colIndex,
        index: colIndex,
        ref: function ref(cell) {
          _this._cells[colIndex] = cell;
        },
        onClick: function onClick() {
          return _this._sortColumn(colIndex);
        }
      }), renderHeader({
        column: column.srcObject,
        columnIndex: colIndex
      }), icon && React.createElement(SortIcon, _extends({}, styleProps, {
        userStyle: userStyle.sort
      }), icon));
    });

    _this.state = {
      columns: _this._formatColumns(props.columns)
    };
    _this._cells = [];
    return _this;
  }

  _createClass(TableHeader, [{
    key: "UNSAFE_componentWillReceiveProps",
    value: function UNSAFE_componentWillReceiveProps(nextProps) {
      if (this.props.columns !== nextProps.columns) {
        this.setState({
          columns: this._formatColumns(nextProps.columns)
        });
      }
    }
  }, {
    key: "_formatColumns",
    value: function _formatColumns(columns) {
      var totalWeight = 0;
      var weights = [];
      columns.forEach(function (col, colIndex) {
        var weight = 1;

        if (colIndex === 0) {
          weight++;
        }

        if (col.type === 'string') {
          weight++;
        }

        weights[colIndex] = weight;
        totalWeight += weight;
      });
      return columns.map(function (col, colIndex) {
        return {
          srcObject: col,
          name: col.name,
          type: col.type,
          sort: SORT.NONE,
          defaultWidth: "".concat(100 / totalWeight * weights[colIndex], "%")
        };
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          theme = _this$props2.theme,
          userStyle = _this$props2.userStyle;
      var columns = this.state.columns;
      return React.createElement(HeaderContainer, {
        theme: theme,
        userStyle: userStyle.header
      }, columns.map(this._renderColumn), React.createElement(AutoSizer, {
        onResize: this._onResize,
        debounceTime: 200
      }));
    }
  }]);

  return TableHeader;
}(PureComponent);

_defineProperty(TableHeader, "propTypes", {
  columns: PropTypes.arrayOf(PropTypes.object),
  renderHeader: PropTypes.func,
  onSort: PropTypes.func,
  onResize: PropTypes.func
});

export { TableHeader as default };
