// @streetscape.gl/monochrome/dist/es5/playback-control/styled-components.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BufferComponent = exports.MarkerComponent = exports.MarkersContainer = exports.TickLabel = exports.Tick = exports.TicksContainer = exports.Timestamp = exports.PlayPauseButton = exports.ControlsContainer = exports.WrapperComponent = void 0;

var _styled = _interopRequireDefault(require("@emotion/styled"));

var _theme = require("../shared/theme");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var WrapperComponent = _styled.default.div(function (props) {
  var result = _objectSpread({}, props.theme.__reset__, {
    boxSizing: 'border-box'
  });

  if (props.isPlaying) {
    result.div = {
      transitionDuration: '0s !important'
    };
  }

  return Object.assign(result, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.WrapperComponent = WrapperComponent;

var ControlsContainer = _styled.default.div(function (props) {
  return _objectSpread({
    display: 'flex',
    alignItems: 'center',
    marginTop: props.theme.spacingTiny
  }, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.ControlsContainer = ControlsContainer;

var PlayPauseButton = _styled.default.div(function (props) {
  return _objectSpread({
    width: 16,
    height: 16,
    marginLeft: props.compact ? 0 : -8,
    marginRight: props.theme.spacingSmall,
    cursor: 'pointer',
    color: props.theme.controlColorPrimary,
    '&:hover': {
      color: props.theme.controlColorHovered
    },
    path: {
      fill: 'currentColor'
    }
  }, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.PlayPauseButton = PlayPauseButton;

var Timestamp = _styled.default.div(function (props) {
  return _objectSpread({}, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.Timestamp = Timestamp;

var TicksContainer = _styled.default.div(function (props) {
  return _objectSpread({
    position: 'relative',
    height: 20
  }, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.TicksContainer = TicksContainer;

var Tick = _styled.default.div(function (props) {
  return _objectSpread({
    height: 4,
    bottom: 0,
    borderLeftStyle: 'solid',
    borderLeftWidth: 1,
    borderLeftColor: props.theme.controlColorSecondary
  }, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.Tick = Tick;

var TickLabel = _styled.default.div(function (props) {
  return _objectSpread({
    transform: 'translate(-50%, -18px)'
  }, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.TickLabel = TickLabel;

var MarkersContainer = _styled.default.div(function (props) {
  return _objectSpread({
    position: 'relative',
    height: 3
  }, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.MarkersContainer = MarkersContainer;

var MarkerComponent = _styled.default.div(function (props) {
  return _objectSpread({
    height: '100%'
  }, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.MarkerComponent = MarkerComponent;

var BufferComponent = _styled.default.div(function (props) {
  return _objectSpread({
    height: '100%',
    background: props.theme.controlColorHovered
  }, (0, _theme.evaluateStyle)(props.userStyle, props));
});

exports.BufferComponent = BufferComponent;
