// @deck.gl/layers/dist/es5/contour-layer/contour-utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.generateContours = generateContours;

var _marchingSquares = require("./marching-squares");

// Given all the cell weights, generates contours for each threshold.

/* eslint-disable max-depth */
function generateContours(_ref) {
  var thresholdData = _ref.thresholdData,
      colors = _ref.colors,
      cellWeights = _ref.cellWeights,
      gridSize = _ref.gridSize,
      gridOrigin = _ref.gridOrigin,
      cellSize = _ref.cellSize;
  var contourSegments = [];
  var contourPolygons = [];
  var width = gridSize[0];
  var height = gridSize[1];
  thresholdData.forEach(function (data, index) {
    var threshold = data.threshold;

    for (var x = -1; x < width; x++) {
      for (var y = -1; y < height; y++) {
        // Get the MarchingSquares code based on neighbor cell weights.
        var _getCode = (0, _marchingSquares.getCode)({
          cellWeights: cellWeights,
          threshold: threshold,
          x: x,
          y: y,
          width: width,
          height: height
        }),
            code = _getCode.code,
            meanCode = _getCode.meanCode;

        var opts = {
          gridOrigin: gridOrigin,
          cellSize: cellSize,
          x: x,
          y: y,
          width: width,
          height: height,
          code: code,
          meanCode: meanCode,
          thresholdData: data
        };

        if (Array.isArray(threshold)) {
          opts.type = _marchingSquares.CONTOUR_TYPE.ISO_BANDS;
          var polygons = (0, _marchingSquares.getVertices)(opts);
          polygons.forEach(function (polygon) {
            contourPolygons.push({
              vertices: polygon,
              threshold: threshold
            });
          });
        } else {
          // Get the intersection vertices based on MarchingSquares code.
          opts.type = _marchingSquares.CONTOUR_TYPE.ISO_LINES;
          var vertices = (0, _marchingSquares.getVertices)(opts);

          for (var i = 0; i < vertices.length; i += 2) {
            contourSegments.push({
              start: vertices[i],
              end: vertices[i + 1],
              threshold: threshold
            });
          }
        }
      }
    }
  });
  return {
    contourSegments: contourSegments,
    contourPolygons: contourPolygons
  };
}
/* eslint-enable max-depth */
