// @loaders.gl/core/dist/esm/binary-utils/binary-utils.js
/* global FileReader, Blob, ArrayBuffer, Buffer */
import assert from '../utils/assert';
import { TextEncoder } from './text-encoding';
export var isArrayBuffer = function isArrayBuffer(x) {
  return x && x instanceof ArrayBuffer;
};
export var isBlob = function isBlob(x) {
  return x && typeof Blob !== 'undefined' && x instanceof Blob;
};
export var isBuffer = function isBuffer(x) {
  return x && x instanceof Buffer;
};
export function toArrayBuffer(data) {
  if (isArrayBuffer(data)) {
    return data;
  } // TODO - per docs we should just be able to call buffer.buffer, but there are issues


  if (isBuffer(data)) {
    var typedArray = new Uint8Array(data);
    return typedArray.buffer;
  } // Careful - Node Buffers will look like ArrayBuffers (keep after isBuffer)


  if (ArrayBuffer.isView(data)) {
    return data.buffer;
  }

  if (typeof data === 'string') {
    var text = data;
    var uint8Array = new TextEncoder().encode(text);
    return uint8Array.buffer;
  }

  return assert(false);
}
export function blobToArrayBuffer(blob) {
  return new Promise(function (resolve, reject) {
    var arrayBuffer;
    var fileReader = new FileReader();

    fileReader.onload = function (event) {
      arrayBuffer = event.target.result;
    };

    fileReader.onloadend = function (event) {
      return resolve(arrayBuffer);
    };

    fileReader.onerror = reject;
    fileReader.readAsArrayBuffer(blob);
  });
} // Convert (copy) ArrayBuffer to Buffer

export function toBuffer(binaryData) {
  if (ArrayBuffer.isView(binaryData)) {
    binaryData = binaryData.buffer;
  }

  if (typeof Buffer !== 'undefined' && binaryData instanceof ArrayBuffer) {
    /* global Buffer */
    var buffer = new Buffer(binaryData.byteLength);
    var view = new Uint8Array(binaryData);

    for (var i = 0; i < buffer.length; ++i) {
      buffer[i] = view[i];
    }

    return buffer;
  }

  return assert(false);
}
export function toDataView(buffer) {
  return new DataView(toArrayBuffer(buffer));
} // Helper functions
