// @deck.gl/core/dist/es6/experimental/utils/gpu-grid-aggregation/gpu-grid-aggregator-utils.js
// Helper methods used by GPUGridAggregator.
import GL from 'luma.gl/constants';
import { Buffer, Framebuffer, Texture2D } from 'luma.gl';
export function getFloatTexture(gl, opts) {
  const _opts$width = opts.width,
        width = _opts$width === void 0 ? 1 : _opts$width,
        _opts$height = opts.height,
        height = _opts$height === void 0 ? 1 : _opts$height;
  const texture = new Texture2D(gl, {
    data: null,
    format: GL.RGBA32F,
    type: GL.FLOAT,
    border: 0,
    mipmaps: false,
    parameters: {
      [GL.TEXTURE_MAG_FILTER]: GL.NEAREST,
      [GL.TEXTURE_MIN_FILTER]: GL.NEAREST
    },
    dataFormat: GL.RGBA,
    width,
    height
  });
  return texture;
}
export function getFramebuffer(gl, opts) {
  const id = opts.id,
        _opts$width2 = opts.width,
        width = _opts$width2 === void 0 ? 1 : _opts$width2,
        _opts$height2 = opts.height,
        height = _opts$height2 === void 0 ? 1 : _opts$height2;
  const texture = opts.texture || getFloatTexture(gl, opts);
  const fb = new Framebuffer(gl, {
    id,
    width,
    height,
    attachments: {
      [GL.COLOR_ATTACHMENT0]: texture
    }
  });
  return fb;
}
export function getFloatArray(array, size) {
  let fillValue = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;

  if (!array || array.length < size) {
    return new Float32Array(size).fill(fillValue);
  }

  return array;
}
export function updateBuffer(_ref) {
  let gl = _ref.gl,
      bufferName = _ref.bufferName,
      data = _ref.data,
      result = _ref.result;

  if (result[bufferName]) {
    result[bufferName].subData({
      data
    });
  } else {
    result[bufferName] = new Buffer(gl, data);
  }
}
