// @streetscape.gl/monochrome/dist/es6/table/styled-components.js
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import styled from '@emotion/styled';
import { evaluateStyle } from '../shared/theme'; // Table

export const WrapperComponent = styled.div(props => _objectSpread({}, props.theme.__reset__, {
  display: 'flex',
  flexDirection: 'column'
}, evaluateStyle(props.userStyle, props)));
export const TableBody = styled.div(props => _objectSpread({
  flex: '1 1 auto'
}, evaluateStyle(props.userStyle, props))); // TableHeader

export const HeaderContainer = styled.div(props => _objectSpread({
  display: 'table',
  flex: '0 0 auto'
}, evaluateStyle(props.userStyle, props)));
export const HeaderCell = styled.div(props => _objectSpread({
  cursor: 'pointer',
  display: 'table-cell',
  position: 'relative',
  fontWeight: 'bold',
  background: props.isAscending || props.isDescending ? props.theme.controlColorActive : props.theme.controlColorPrimary,
  color: props.theme.textColorInvert,
  paddingTop: props.theme.spacingTiny,
  paddingBottom: props.theme.spacingTiny,
  paddingLeft: props.theme.spacingSmall,
  paddingRight: props.theme.spacingSmall,
  '&:hover': {
    background: props.theme.controlColorHovered
  }
}, evaluateStyle(props.userStyle, props)));
export const SortIcon = styled.div(props => _objectSpread({
  position: 'absolute',
  right: props.theme.spacingTiny,
  bottom: props.theme.spacingTiny
}, evaluateStyle(props.userStyle, props)));
export const TableRowComponent = styled.div(props => _objectSpread({
  position: 'relative',
  display: 'flex',
  alignItems: 'stretch',
  borderBottomStyle: 'solid',
  borderBottomWidth: 1,
  borderBottomColor: props.theme.controlColorSecondary
}, evaluateStyle(props.userStyle, props)));
export const TableCell = styled.div(props => _objectSpread({
  flex: '0 0 auto',
  overflow: 'hidden',
  boxSizing: 'border-box',
  position: 'relative',
  paddingTop: props.theme.spacingTiny,
  paddingBottom: props.theme.spacingTiny,
  paddingLeft: props.theme.spacingSmall,
  paddingRight: props.theme.spacingSmall,
  borderLeftStyle: 'solid',
  borderLeftWidth: props.index === 0 ? 0 : 1,
  borderLeftColor: props.theme.controlColorSecondary
}, evaluateStyle(props.userStyle, props)));
export const Expander = styled.div(props => _objectSpread({
  position: 'absolute',
  cursor: 'pointer',
  left: -props.theme.spacingSmall
}, evaluateStyle(props.userStyle, props)));
