// @streetscape.gl/monochrome/dist/es6/table/table.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import AutoSizer from 'react-virtualized/dist/commonjs/AutoSizer';
import List from 'react-virtualized/dist/commonjs/List';
import CellMeasurer, { CellMeasurerCache } from 'react-virtualized/dist/commonjs/CellMeasurer';
import { withTheme } from '../shared/theme';
import TableHeader from './table-header';
import TableRow from './table-row';
import { WrapperComponent, TableBody } from './styled-components';
export class Table extends PureComponent {
  constructor(props) {
    super(props);

    _defineProperty(this, "_onHeaderResize", columns => {
      this.setState({
        columns
      }, this._forceUpdate);
    });

    _defineProperty(this, "_onSort", sortFunc => {
      const rows = this.state.rows;

      if (sortFunc) {
        rows.sort(sortFunc);
      }

      this.setState({
        sortFunc
      });

      this._forceUpdate();
    });

    _defineProperty(this, "_forceUpdate", () => {
      if (this._list) {
        this._cache.clearAll();

        this._list.recomputeRowHeights();
      }
    });

    _defineProperty(this, "_renderRowMeasurer", (_ref) => {
      let key = _ref.key,
          parent = _ref.parent,
          index = _ref.index,
          style = _ref.style;
      return React.createElement(CellMeasurer, {
        cache: this._cache,
        parent: parent,
        key: key,
        rowIndex: index,
        collumnIndex: 0
      }, () => this._renderRow({
        key,
        index,
        style
      }));
    });

    this.state = {
      columns: null,
      sortFunc: null,
      rows: this._formatRows(props.rows)
    };
    this._cache = new CellMeasurerCache({
      fixedWidth: true
    });
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    if (this.props.rows !== nextProps.rows) {
      this.setState({
        rows: this._formatRows(nextProps.rows, this.state.sortFunc)
      });

      this._forceUpdate();
    }
  }

  _formatRows(rows, sortFunc) {
    rows = rows.map((row, rowIndex) => ({
      srcObject: row,
      data: row.data,
      children: row.children,
      id: String(rowIndex)
    }));

    if (sortFunc) {
      rows.sort(sortFunc);
    }

    return rows;
  }

  _renderRow(_ref2) {
    let key = _ref2.key,
        index = _ref2.index,
        style = _ref2.style;
    const _this$props = this.props,
          renderCell = _this$props.renderCell,
          theme = _this$props.theme,
          userStyle = _this$props.style;
    const row = this.state.rows[index];
    return React.createElement(TableRow, {
      key: key,
      id: row.id,
      index: index,
      data: row,
      style: style,
      theme: theme,
      userStyle: userStyle,
      renderCell: renderCell,
      columns: this.state.columns
    });
  }

  // AutoSizer is a pure component. By default child function is only called if dimensions change.
  // Rebind this function every render to respond to state change
  _renderBody(_ref3) {
    let width = _ref3.width,
        height = _ref3.height;
    const _this$state = this.state,
          columns = _this$state.columns,
          rows = _this$state.rows;

    if (!columns) {
      return null;
    }

    return React.createElement(List, {
      ref: list => {
        this._list = list;
      },
      tabIndex: null,
      height: height,
      rowCount: rows.length,
      rowHeight: this._cache.rowHeight,
      rowRenderer: this._renderRowMeasurer,
      width: width
    });
  }

  render() {
    const _this$props2 = this.props,
          theme = _this$props2.theme,
          width = _this$props2.width,
          height = _this$props2.height,
          style = _this$props2.style,
          columns = _this$props2.columns,
          renderHeader = _this$props2.renderHeader;
    return React.createElement(WrapperComponent, {
      style: {
        width,
        height
      },
      theme: theme,
      userStyle: style.wrapper
    }, React.createElement(TableHeader, {
      theme: theme,
      userStyle: style,
      columns: columns,
      renderHeader: renderHeader,
      onSort: this._onSort,
      onResize: this._onHeaderResize
    }), React.createElement(TableBody, {
      theme: theme,
      userStyle: style.body
    }, React.createElement(AutoSizer, null, this._renderBody.bind(this))));
  }

}

_defineProperty(Table, "propTypes", {
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  style: PropTypes.object,
  columns: PropTypes.array,
  rows: PropTypes.array,
  renderHeader: PropTypes.func,
  renderCell: PropTypes.func
});

_defineProperty(Table, "defaultProps", {
  width: '100%',
  height: 400,
  style: {},
  rows: [],
  renderHeader: (_ref4) => {
    let column = _ref4.column;
    return column.name;
  },
  renderCell: (_ref5) => {
    let value = _ref5.value;
    return value === null ? null : String(value);
  }
});

export default withTheme(Table);
