// @streetscape.gl/monochrome/dist/es6/table/tree-table-row.js
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { TableRowComponent, TableCell, Expander } from './styled-components';
/**
 * A stateless component that renders a data row in the TreeTable component
 */

export default class TreeTableRow extends PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "_renderItem", (_ref) => {
      let id = _ref.id,
          index = _ref.index,
          _ref$depth = _ref.depth,
          depth = _ref$depth === void 0 ? 0 : _ref$depth,
          key = _ref.key,
          data = _ref.data,
          style = _ref.style;
      const _this$props = this.props,
            indentSize = _this$props.indentSize,
            columns = _this$props.columns,
            theme = _this$props.theme,
            userStyle = _this$props.userStyle,
            renderCell = _this$props.renderCell,
            getIsExpanded = _this$props.getIsExpanded;
      const isExpanded = getIsExpanded(id);
      const hasChildren = data.children.length > 0;
      const indent = indentSize * (depth + 1);
      return React.createElement("div", {
        key: key,
        style: style
      }, React.createElement(TableRowComponent, {
        theme: theme,
        index: index,
        userStyle: userStyle.row
      }, hasChildren && React.createElement(Expander, {
        key: "toggle",
        isExpanded: isExpanded,
        theme: theme,
        userStyle: userStyle.expander,
        style: {
          marginLeft: indent
        },
        onClick: () => this._toggleExpansion(id)
      }, isExpanded ? userStyle.iconExpanded || '▾' : userStyle.iconCollapsed || '▸'), React.createElement("div", {
        style: {
          flexShrink: 0,
          width: indent
        }
      }), data.data.map((colValue, colIndex) => {
        const column = columns[colIndex];
        const width = colIndex === 0 ? column.width - indent : column.width;
        return React.createElement(TableCell, {
          key: colIndex,
          index: colIndex,
          style: {
            width
          },
          title: `${column.name}: ${colValue}`,
          theme: theme,
          userStyle: userStyle.cell
        }, renderCell({
          value: colValue,
          column: column.srcObject,
          columnIndex: colIndex,
          row: data.srcObject,
          rowId: id
        }));
      })), hasChildren && isExpanded && data.children.map((row, rowIndex) => this._renderItem({
        depth: depth + 1,
        id: `${id}.${rowIndex}`,
        index: rowIndex,
        key: rowIndex,
        data: row
      })));
    });
  }

  _toggleExpansion(id) {
    this.props.toggleExpansion(id, this.props.id);
  }

  render() {
    return this._renderItem(_objectSpread({}, this.props));
  }

}

_defineProperty(TreeTableRow, "propTypes", {
  id: PropTypes.string,
  data: PropTypes.shape({
    data: PropTypes.array,
    children: PropTypes.array
  }),
  style: PropTypes.object,
  indentSize: PropTypes.number,
  renderCell: PropTypes.func,
  columns: PropTypes.array,
  getIsExpanded: PropTypes.func,
  toggleExpansion: PropTypes.func
});
