// @xviz/parser/dist/esm/parsers/parse-log-metadata.js
function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { getXVIZConfig, setXVIZConfig } from '../config/xviz-config';
import { parseVersionString } from './xviz-v2-common'; // Post-processes log metadata

export function parseLogMetadata(data) {
  var _getXVIZConfig = getXVIZConfig(),
      supportedVersions = _getXVIZConfig.supportedVersions;

  var versionString = data.version;
  var currentMajorVersion = null;

  if (versionString === undefined) {
    currentMajorVersion = 1;
  } else {
    var _parseVersionString = parseVersionString(versionString),
        major = _parseVersionString.major;

    currentMajorVersion = major;
  }

  if (!currentMajorVersion) {
    throw new Error('Unable to detect the XVIZ version.');
  } else {
    setXVIZConfig({
      currentMajorVersion: currentMajorVersion
    });
  }

  if (supportedVersions && !supportedVersions.includes(currentMajorVersion)) {
    throw new Error("XVIZ version ".concat(currentMajorVersion, " is not supported.  Currently supported versions are ").concat(supportedVersions, "."));
  }

  return currentMajorVersion === 1 ? parseLogMetadataV1(data) : parseLogMetadataV2(data);
}
export function parseLogMetadataV1(data) {
  // streams is the map from stream names (ie streams) to the url resource
  var originalStreams = data.streams || []; // Use XVIZ configuration to filter out unwanted / blacklisted streams

  var _getXVIZConfig2 = getXVIZConfig(),
      STREAM_BLACKLIST = _getXVIZConfig2.STREAM_BLACKLIST;

  var streams = {};
  Object.keys(originalStreams).forEach(function (streamName) {
    if (!STREAM_BLACKLIST.has(streamName)) {
      streams[streamName] = originalStreams[streamName];
    }
  });

  var _getTimestamps = getTimestamps(data),
      logStartTime = _getTimestamps.logStartTime,
      logEndTime = _getTimestamps.logEndTime,
      eventStartTime = _getTimestamps.eventStartTime,
      eventEndTime = _getTimestamps.eventEndTime;

  var metadata = _objectSpread({}, data, {
    streams: streams,
    logStartTime: logStartTime,
    logEndTime: logEndTime,
    eventStartTime: eventStartTime,
    eventEndTime: eventEndTime
  });

  return metadata;
}
export function parseLogMetadataV2(data) {
  // streams is the map from stream names (ie streams) to the url resource
  var originalStreams = data.streams; // Use XVIZ configuration to filter out unwanted / blacklisted streams

  var _getXVIZConfig3 = getXVIZConfig(),
      STREAM_BLACKLIST = _getXVIZConfig3.STREAM_BLACKLIST;

  var streams = {};

  if (originalStreams) {
    Object.keys(originalStreams).forEach(function (streamName) {
      if (!STREAM_BLACKLIST.has(streamName)) {
        streams[streamName] = originalStreams[streamName];
      }
    });
  }

  var logInfo = data.log_info || {};

  var _getTimestamps2 = getTimestamps(logInfo),
      logStartTime = _getTimestamps2.logStartTime,
      logEndTime = _getTimestamps2.logEndTime,
      eventStartTime = _getTimestamps2.eventStartTime,
      eventEndTime = _getTimestamps2.eventEndTime;

  var styles = collectStreamStyles(streams);

  var metadata = _objectSpread({}, data, {
    streams: streams,
    // Overrides entry from 'data'
    logStartTime: logStartTime,
    logEndTime: logEndTime,
    start_time: eventStartTime,
    // eslint-disable-line camelcase
    end_time: eventEndTime,
    // eslint-disable-line camelcase
    // TODO: i don't think these are ever used
    eventStartTime: eventStartTime,
    eventEndTime: eventEndTime,
    styles: styles
  });

  return metadata;
}

function getTimestamps(info) {
  var logStartTime = Number.isFinite(info.log_start_time) ? info.log_start_time : null;
  var logEndTime = Number.isFinite(info.log_end_time) ? info.log_end_time : null; // Fallback to complete log time if we don't have a specific playback time range

  var eventStartTime = Number.isFinite(info.start_time) ? info.start_time : logStartTime;
  var eventEndTime = Number.isFinite(info.end_time) ? info.end_time : logEndTime;
  return {
    logStartTime: logStartTime,
    logEndTime: logEndTime,
    eventStartTime: eventStartTime,
    eventEndTime: eventEndTime
  };
}
/**
 * Convert V2 stylesheet data to the internal representation
 */


function collectStreamStyles(metadataStreams) {
  var internalStylesheet = {};
  Object.keys(metadataStreams).forEach(function (streamId) {
    var streamMetadata = metadataStreams[streamId];
    var streamStylesheet = [];

    if (streamMetadata.stream_style) {
      streamStylesheet.push({
        name: '*',
        style: streamMetadata.stream_style
      });
    }

    if (streamMetadata.style_classes) {
      streamStylesheet.push.apply(streamStylesheet, _toConsumableArray(streamMetadata.style_classes));
    }

    if (streamStylesheet.length !== 0) {
      internalStylesheet[streamId] = streamStylesheet;
    }
  });
  return internalStylesheet;
}
