// @xviz/builder/dist/es5/writers/xviz-writer/xviz-writer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _xvizBinaryWriter = require("./xviz-binary-writer");

var _xvizJsonEncoder = require("./xviz-json-encoder.js");

var _draco = require("@loaders.gl/draco");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

// 0-frame is an index file for timestamp metadata
// 1-frame is the metadata file for the log
// 2-frame is where the actual XVIZ updates begin
var frameName = function frameName(index) {
  return "".concat(index + 2, "-frame");
};
/**
 * Class to abstract away file IO
 */


var FileSink =
/*#__PURE__*/
function () {
  function FileSink() {
    _classCallCheck(this, FileSink);

    this.fs = module.require('fs');
    this.path = module.require('path');
  }

  _createClass(FileSink, [{
    key: "writeSync",
    value: function writeSync(scope, name, data) {
      var xvizMetadataFilename = this.path.join(scope, name);
      var options = {
        flag: 'w'
      };
      this.fs.writeFileSync(xvizMetadataFilename, data, options);
    }
  }]);

  return FileSink;
}();

var XVIZWriter =
/*#__PURE__*/
function () {
  function XVIZWriter() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    _classCallCheck(this, XVIZWriter);

    var _options$dataSink = options.dataSink,
        dataSink = _options$dataSink === void 0 ? new FileSink() : _options$dataSink,
        _options$envelope = options.envelope,
        envelope = _options$envelope === void 0 ? true : _options$envelope,
        _options$binary = options.binary,
        binary = _options$binary === void 0 ? true : _options$binary,
        _options$json = options.json,
        json = _options$json === void 0 ? false : _options$json,
        _options$draco = options.draco,
        draco = _options$draco === void 0 ? false : _options$draco;
    this.sink = dataSink;
    this.frameTimings = {
      frames: new Map()
    };
    this.wroteFrameIndex = null;
    this.options = {
      envelope: envelope,
      binary: binary,
      json: json,
      draco: draco
    };
  } // xvizMetadata is the object returned
  // from a Builder.


  _createClass(XVIZWriter, [{
    key: "writeMetadata",
    value: function writeMetadata(xvizDirectory, xvizMetadata) {
      this._saveTimestamp(xvizMetadata);

      if (this.options.envelope) {
        xvizMetadata = {
          type: 'xviz/metadata',
          data: xvizMetadata
        };
      }

      if (this.options.binary) {
        var options = {
          flattenArrays: false
        };
        (0, _xvizBinaryWriter.writeBinaryXVIZtoFile)(this.sink, xvizDirectory, '1-frame', xvizMetadata, options);
      }

      if (this.options.json) {
        this.sink.writeSync(xvizDirectory, '1-frame.json', JSON.stringify(xvizMetadata));
      }
    }
  }, {
    key: "writeFrame",
    value: function writeFrame(xvizDirectory, frameIndex, xvizFrame) {
      if (this.wroteFrameIndex !== null) {
        throw new Error("writeFrame() was called after writeFrameIndex().  The index was written with last frame of ".concat(frameName(this.wroteFrameIndex - 1)));
      }

      this._saveTimestamp(xvizFrame, frameIndex);

      if (this.options.envelope) {
        xvizFrame = {
          type: 'xviz/state_update',
          data: xvizFrame
        };
      }

      if (this.options.binary) {
        var options = {
          flattenArrays: false
        };

        if (this.options.draco) {
          options.DracoEncoder = _draco.DracoEncoder;
          options.DracoDecoder = _draco.DracoDecoder;
        }

        (0, _xvizBinaryWriter.writeBinaryXVIZtoFile)(this.sink, xvizDirectory, frameName(frameIndex), xvizFrame, options);
      }

      if (this.options.json) {
        // Limit precision to save space
        var numberRounder = function numberRounder(k, value) {
          if (typeof value === 'number') {
            return Number(value.toFixed(10));
          }

          return value;
        };

        var jsonXVIZFrame = (0, _xvizJsonEncoder.xvizConvertJson)(xvizFrame);
        this.sink.writeSync(xvizDirectory, "".concat(frameName(frameIndex), ".json"), JSON.stringify(jsonXVIZFrame, numberRounder));
      }
    }
  }, {
    key: "writeFrameIndex",
    value: function writeFrameIndex(xvizDirectory) {
      var _this$frameTimings = this.frameTimings,
          startTime = _this$frameTimings.startTime,
          endTime = _this$frameTimings.endTime,
          frames = _this$frameTimings.frames;
      var frameTimings = {};

      if (startTime) {
        frameTimings.startTime = startTime;
      }

      if (endTime) {
        frameTimings.endTime = endTime;
      } // Sort frames by index before writing out as an array


      var frameTimes = Array.from(frames.keys()).sort(function (a, b) {
        return a - b;
      });
      var timing = [];
      frameTimes.forEach(function (value, index) {
        // Value is two greater than frame index
        var limit = timing.length;

        if (value > limit) {
          // Adding 2 because 1-frame is metadata file, so frame data starts at 2
          throw new Error("Error writing time index file. Frames are missing between ".concat(limit + 2, " and ").concat(value + 2));
        }

        timing.push(frames.get(value));
      });
      frameTimings.timing = timing;
      this.sink.writeSync(xvizDirectory, '0-frame.json', JSON.stringify(frameTimings));
      this.wroteFrameIndex = timing.length;
    }
    /* eslint-disable camelcase */

  }, {
    key: "_saveTimestamp",
    value: function _saveTimestamp(xviz_data, index) {
      var log_info = xviz_data.log_info,
          updates = xviz_data.updates;

      if (index === undefined) {
        // Metadata case
        if (log_info) {
          var _ref = log_info || {},
              start_time = _ref.start_time,
              end_time = _ref.end_time;

          if (start_time) {
            this.frameTimings.startTime = start_time;
          }

          if (end_time) {
            this.frameTimings.endTime = end_time;
          }
        }
      } else if (updates) {
        if (updates.length === 0 || !updates.every(function (update) {
          return typeof update.timestamp === 'number';
        })) {
          throw new Error('XVIZ updates did not contain a valid timestamp');
        }

        var min = Math.min(updates.map(function (update) {
          return update.timestamp;
        }));
        var max = Math.max(updates.map(function (update) {
          return update.timestamp;
        }));
        this.frameTimings.frames.set(index, [min, max, index, frameName(index)]);
      } else {
        // Missing updates & index is invalid call
        throw new Error('Cannot find timestamp');
      }
    }
    /* eslint-enable camelcase */

  }]);

  return XVIZWriter;
}();

exports.default = XVIZWriter;
