// @loaders.gl/core/dist/es5/read-file/read-file.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.readFile = readFile;
exports.readFileSync = readFileSync;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _fs = _interopRequireDefault(require("fs"));

var _http = _interopRequireDefault(require("http"));

var _https = _interopRequireDefault(require("https"));

var _util = _interopRequireDefault(require("util"));

var _fileAliases = require("./file-aliases");

var _decodeDataUri = _interopRequireDefault(require("../data-uri-utils/decode-data-uri"));

var _binaryUtils = require("../binary-utils/binary-utils");

var _streamUtils = require("../async-iterator-utils/stream-utils");

/* global fetch */

/* global URL, location, File, FileReader */

/* global Buffer */
// `fs` will be empty object in browsers (see package.json "browser" field).
var isNode = Boolean(_fs.default && _fs.default.readFile);
var DEFAULT_OPTIONS = {
  dataType: 'arrayBuffer'
};

function getReadFileOptions() {
  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  options = Object.assign({}, DEFAULT_OPTIONS, options);
  options.responseType = options.responseType || options.dataType;

  if (_fs.default) {
    // set encoding for fs.readFile
    options.encoding = options.encoding || (options.dataType === 'text' ? 'utf8' : null);
  }

  return options;
} // Reads raw file data from:
// * http/http urls
// * data urls
// * File/Blob objects
// etc?


function readFile(uri) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  try {
    uri = (0, _fileAliases.resolvePath)(uri);
    options = getReadFileOptions(options);

    if (uri.startsWith('data:')) {
      return Promise.resolve((0, _decodeDataUri.default)(uri));
    }

    if (typeof File !== 'undefined' && uri instanceof File) {
      readFileObject(uri, options);
    }

    if (isNode) {
      var isRequest = uri.startsWith('http:') || uri.startsWith('https:');

      if (isRequest) {
        return new Promise(function (resolve, reject) {
          options = (0, _objectSpread2.default)({}, new URL(uri), options);
          var request = uri.startsWith('https:') ? _https.default.request : _http.default.request;
          request(uri, function (response) {
            return (0, _streamUtils.concatenateReadStream)(response).then(resolve, reject);
          });
        });
      }

      return readFileNode(uri, options);
    } // Browser: Try to load all URLS via fetch, as they can be local requests (e.g. to a dev server)
    // In a web worker, XMLHttpRequest throws invalid URL error if using relative path
    // resolve url relative to original base


    uri = new URL(uri, location.href).href;
    return fetch(uri, options).then(function (res) {
      return res[options.dataType]();
    }); // return Promise.reject(new Error('Cannot load file URIs in browser'));
  } catch (error) {
    return Promise.reject(error.message);
  }
} // In a few cases (data URIs, node.js) "files" can be read synchronously


function readFileSync(uri) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  uri = (0, _fileAliases.resolvePath)(uri);
  options = getReadFileOptions(options);

  if (uri.startsWith('data:')) {
    return (0, _decodeDataUri.default)(uri);
  }

  if (!isNode) {
    return null; // throw new Error('Cant load URI synchronously');
  }

  var buffer = _fs.default.readFileSync(uri, options, function () {});

  return buffer instanceof Buffer ? (0, _binaryUtils.toArrayBuffer)(buffer) : buffer;
} // HELPERS

/**
 * File reader function for the browser
 * @param {File|Blob} file  HTML File or Blob object to read as string
 * @returns {Promise.string}  Resolves to a string containing file contents
 */


function readFileObject(file, options) {
  return new Promise(function (resolve, reject) {
    try {
      var reader = new FileReader();

      reader.onerror = function (error) {
        return reject(new Error(error));
      };

      reader.onabort = function () {
        return reject(new Error('Read operation was aborted.'));
      };

      reader.onload = function () {
        return resolve(reader.result);
      };

      if (options.dataType !== 'arraybuffer') {
        reader.readAsText(file);
      } else {
        reader.readAsArrayBuffer(file);
      }
    } catch (error) {
      reject(error);
    }
  });
}

function readFileNode(filename, options) {
  var readFileAsync = _util.default.promisify(_fs.default.readFile);

  return readFileAsync(filename, options).then(function (buffer) {
    return buffer instanceof Buffer ? (0, _binaryUtils.toArrayBuffer)(buffer) : buffer;
  });
}
